First draft of the extension to VDA 5050 with the zone concept (#10)
ConstantinUp/VDA5050: VDA5050_EN.md

Reviewer: 6.x Zones: 
First paragraph, second to last sentence. Blocked -> 'BLOCKED'

Reviewer: 6.x Zones: 
Last paragraph zoneSet -> zone set

Reviewer: 6.x.1 Zone types: First paragraph
-> the behaviour that defines zone entry/ exit? Not the one that is observed during entry/exit?

Reviewer: 6.x.1 Zone types
Contour-based zones: RELEASE has "none" as parameter in table

Reviewer: 6.x.1 Zone types: First paragraph
-> the behaviour that defines zone entry/ exit? Not the one that is observed during entry/exit?

Reviewer: 6.x.1 Zone types
Contour-based zones: ACTION has "Change:" in description, should be removed

Reviewer: 6.x.1 Zone types:
Kinematic center-based zones: 
First sentence has "but have" twice

Reviewer: PRIORITY and PENALTY zones represent zones, that are **used** for the path planning of the robots, but have, but have no effect on the behaviour of the robots when passing through. --> I propose to change this to **must be respected** / **influence the robots navigation/ path planning**

Reviewer: A PRIORITY ... to an otherwise equivalent area with no zone on the map.  --> A PRIORITY ... to an otherwise equivalent area with no **such** zone on the map. 

Same for " A PENALTY ... otherwise equivalent area with no zone on the map."
-> there might also be other zones on the map already

Reviewer: With a maximum penaltyFactor of 1.0, vehicles shall avoid this zone when possible and only use it if no other cheaper way exists. n this this case, the zone is not yet fully restricted as a BLOCKED zone, but should be avoided.
-->

With a maximum penaltyFactor of 1.0, vehicles shall avoid this zone **removed** and only use it if no other **feasible** way exists. **In** this case, the zone is not yet fully restricted **like** a BLOCKED zone but should be avoided **at all costs.**

Reviewer: Figure x2: "navigation-point" used instead of "kinematic-centre"

Reviewer: Table of kinematic-centre-based zones: 
DIRECTED:
OBSTACLE: Vehicles may not traverse against the direction of travel except to avoid obstacles
--> OBSTACLE: Vehicles may **deviate from the preferred direction of travel** except to avoid obstacles, **but never travel in the opposite direction**

Author: To be discussed in meeting

Reviewer: Table of kinematic-centre-based zones:
DIRECTED:

STRICT: Vehicles shall not traverse against the direction of travel within the zone.
--> Vehicles shall not deviate from the preferred direction of travel within this zone.

Reviewer: Proposal to make these definitions consistent with the definitions for the TWOWAYS zone?

Reviewer: Would probably move this part: 
6.x.2 Sharing of planned path for freely navigating mobile robots
at the very beginning of the section 6.2

Reviewer: In the definition of a zone object: 

<!DOCTYPE html><ins style="box-sizing: border-box; box-shadow: inset 4px 0 0 var(--borderColor-success-muted, var(--color-success-muted)); display: block; border-radius: 0px; border-top-width: 0px; border-top-style: none; border-top-color: currentcolor; border-bottom-width: 0px; border-bottom-style: none; border-bottom-color: currentcolor; text-decoration: none; caret-color: rgb(240, 246, 252); color: rgb(240, 246, 252); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 16px; font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-stroke-width: 0px;"><p dir="auto" class="rich-diff-level-zero" style="box-sizing: border-box; margin-top: 0px; margin-bottom: var(--base-size-16); margin-left: var(--base-size-16);">A single zone object has the following structure:</p></ins><ins style="box-sizing: border-box; box-shadow: inset 4px 0 0 var(--borderColor-success-muted, var(--color-success-muted)); display: block; border-radius: 0px; border-top-width: 0px; border-top-style: none; border-top-color: currentcolor; border-bottom-width: 0px; border-bottom-style: none; border-bottom-color: currentcolor; text-decoration: none; caret-color: rgb(240, 246, 252); color: rgb(240, 246, 252); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 16px; font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-stroke-width: 0px;"><markdown-accessiblity-table class="rich-diff-level-zero" data-catalyst="" style="box-sizing: border-box; display: block; margin-left: var(--base-size-16);">
Object structure | Data type | Description
-- | -- | --
zone{ | JSON object |  
zoneId | string | Locally (within the zone set) unique identifier.
zoneType | string | Zone category.

</markdown-accessiblity-table></ins>

zoneType should be enum

Reviewer: <!DOCTYPE html>
vertices[vertex]} | array | Array of vertices (in x-y-coordinates) defining the geometrical shape of the zone.
-- | -- | --

no closing bracket after the vertices. Also should we first list all parameters and then the vertices defining the polygon?

Reviewer: <!DOCTYPE html>
releaseLossBehavior} | enum | When the access to a releaseZone is revoked or expired, the vehicle can either STOP, CONTINUE, or EVACUATE the zone. This Action is only executed, when the vehicle is already in the zone and the release expires or is revoked. If not defined, the vehicle is expected to STOP and report an error. STOP: Vehicle stops, sends an fatal error. EVACUATE: Execute the evacuation behavior of the vehicle to leave the zone, keeping the release in its state until the zone is left. CONTINUE: If the revoke/expiry happens, after the vehicle entered the zone, the vehicle continue its path keeping the reservation of the vehicle in its state. If the order ends inside the zone, the vehicle waits for new order.
-- | -- | --

Here "CONTINUE" is still listed within the ENUM; but not at the very beginning of 6.x

Author: To be discussed in tomorrow's meeting

Reviewer: The array of vertices defining a zone is provided as a list of x-y-tuples given in the globally defined project specific coordinate system:

-> definition of the polygon in clockwise order? Did we not want to add that as some libraries require this as a parameter when called?

Author: You are correct, clockwise was added. Presented in tomorrows meeting.

Reviewer: For communicating and verifying requests for interactive zones, the state message is used by the vehicle to request an authorization and a the MQTT topic zones is used by master control to grant or revoke the authorization. The authorization request is related to the vehicle state in which the request is embedded.

-> did we not wish to have a dedicated /response topic? why is the request answered in the /zones topic?

Author: should be the /response topic

Reviewer: The **boolean** requestStatus shall be initially set to REQUESTED by the vehicle for stating its request. If master control grants the request (as described in the following section) and the vehicle receives the corresponding message, it acknowledges that the vehicle has received the permission, by setting the authorized field to true. The vehicle has now the permission to enter the RELEASE zone.
-> why should the requestStatus be BOOLEAN?

The master control can request the deletion of a specific map from a vehicle. This is done with the instant action `deleteMap`. When a vehicle runs out of memory, it should report this to the master control, which can then initiate the deletion of maps. The vehicle itself is not allowed to delete maps.
After successfully deleting a map, it is important to remove that map's entry from the vehicle's array of maps in the vehicle state.

## 6.2 Sharing of planned path for freely navigating mobile robots (Standalone chapter)

> Vehicles shall communicate their planned trajectory to the master control system. This is done via the state message. For a higher frequency of sharing, the visualization topic can be used.

Reviewer 2: ```suggestion
Vehicles shall communicate their planned trajectory to the master control system. This is done via the state message. For a higher frequency of sharing, the `visualization` topic can be used.
```

The parameters `plannedPath` and `intermediatePath` shall be used only for trajectories planned by the mobile robot. The trajectory fields in the edgeState shall only be used to 'acknowledge' trajectories that have already been defined a priori within a layout or the order.

## 6.x Zones

Zones are used to define rules for specific areas of the vehicle workspace. In this way, zones allow vehicles to navigate freely between nodes while giving the Master Control the ability to manage traffic. Zones can be used to locally deny vehicles access to areas or to link access to conditions (zone types: 'BLOCKED' and 'RELEASE'). It is also possible to enforce specific behavior while within the zone (zone types: 'LINE_GUIDED', 'SPEED_LIMIT', 'COORDINATED_REPLANNING', and 'ACTION') or influence the driving behavior by incentivizing or penalizing certain areas (zone types: 'PRIORITY' and 'PENALTY') or giving a predefined driving direction (zone type: 'DIRECTED', 'BIDIRECTED'). The zone types are defined in the following sections.

> Potential conflicts in orders due to overlapping of zones or combination of zone and edge properties and how to resolve them are addressed in section 10.4.

Reviewer: 6.x?

Author: ```suggestion
Potential conflicts in orders due to overlapping of zones or combination of zone and edge properties and how to resolve them are addressed in section 6.x.
```

Some vehicles cannot process zones at all, while other vehicles might only be able to work with a certain subset of zone types, such as 'BLOCKED'. All vehicles must therefore report to master control which zones they are able to understand by adding the according zone names to the `supportedZones` array under `typeSpecifications` in their factsheet.
Also (virtually) line-guided vehicles can choose to support zone-based navigation if they can implement the logic of the corresponding zone types defined in the following. 

A zone set shall only be changed and distributed by master control to keep consistency in the system.

### 6.x.1 Zone types

> Two types of zones are distinguished: contour-based zones and kinematic center-based zones. This distinction is based on the different conditions for when the vehicle is considered to be entering and exiting zones.

Reviewer: Two categories (or groups)... zone type is already a term used

Author: ```suggestion
Two categories of zones are distinguished: contour-based zones and kinematic center-based zones. This distinction is based on the different conditions for when the vehicle is considered to be entering and exiting zones.
```

#### Contour-based zones

> For contour-based zones, the contour of the vehicle (including its load) determines zone entry and exit. Any part of the contour entering the zone is considered a zone entry. A zone exit shall be reported only when no part of the contour is inside the zone. When a zone is reported as exited is decided by the mobile robot. 

Reviewer: A zone exit shall be reported only when no part of the contour remains inside the zone.

Author: ```suggestion
For contour-based zones, the contour of the vehicle (including its load) determines zone entry and exit. Any part of the contour entering the zone is considered a zone entry. A zone exit shall be reported only when no part of the contour remains inside the zone. When a zone is reported as exited is decided by the mobile robot. 
```

![Figure x1 Depiction of a vehicle entering a zone based on its contour (left) and a loaded vehicle with corresponding extended bounding box exiting a zone (right)](./assets/contour_entry.png)
>Figure x1 Depiction of a vehicle entering a zone based on its contour (left) and a loaded vehicle with corresponding extended bounding box exiting a zone (right)

The following contour-based zones are defined:

| **Zone Type**| **Zone Parameters** | **Data type** | **Description** | 
| --- | --- | --- | --- |

> | BLOCKED | none | | Vehicles shall not enter this zone. If a vehicle has entered the zone or finds itself within one, it shall stop and throw a FATAL error.| 

Reviewer: CRTICAL error?

Author: We discussed that in the last meeting, since I was not sure about this one, but the feedback from the room was to still have a FATAL error. The argument was, that BLOCKED zones are not to be entered at all and MC does not have the authority to decide to change that. Counter-argument would be a vehicle that is within the zone through a localization error and is recovering afterwards again. Move discussion to next meeting.

Author: ```suggestion
| BLOCKED | none | | Vehicles shall not enter this zone. If a vehicle has entered the zone or finds itself within one, it shall stop and throw a CRITICAL error.| 
```

| LINE_GUIDED | none | | No free navigation is allowed in this zone, mobile robots shall follow the defined trajectories on edges. Vehicles may only enter this zone if the route is explicitly specified by the master control in the form of a node-edge graph. Any movement of the mobile robot that requires it to enter this zone shall follow a predefined trajectory. When entering the zone, the mobile robot shall be on the trajectory of the edge that crosses the zone. The edges that enter and are inside the line-guided zone require a trajectory sent from the master control or a predefined trajectory on the vehicle. A corridor can be sent to allow the vehicle to deviate from the trajectory. | 
| RELEASE | | - | Vehicles are only allowed entering this zone once they have been granted access through master control. | 

> | | releaseLossBehavior | string | Enum {'STOP', 'CONTINUE', 'EVACUATE'}</br>When the access to this zone is revoked or expired, the vehicle can either STOP, CONTINUE, or EVACUATE the zone. This action is only executed, when the vehicle is already in the zone and the release expires or is revoked. If not defined, the vehicle is expected to STOP and report an error. STOP: Vehicle stops and sends a CRITICAL error. EVACUATE: Execute the evacuation behavior of the vehicle to leave the zone, keeping the release in its state until the zone is left. CONTINUE: If the revoke/expiry happens, after the vehicle entered the zone, the vehicle continues its path keeping the release of the zone in its state. If the order ends inside the zone, the vehicle waits for a new order. | 

Reviewer: EVACUATE: Execute the evacuation behavior of the vehicle to leave the zone, keeping the `zoneRequest` object granting release in its state until the zone is left.

Author: ```suggestion
| | releaseLossBehavior | string | Enum {'STOP', 'CONTINUE', 'EVACUATE'}</br>When the access to this zone is revoked or expired, the vehicle can either STOP, CONTINUE, or EVACUATE the zone. This action is only executed, when the vehicle is already in the zone and the release expires or is revoked. If not defined, the vehicle is expected to STOP and report an error. STOP: Vehicle stops and sends a CRITICAL error. EVACUATE: Execute the evacuation behavior of the vehicle to leave the zone, keeping the `zoneRequest` object granting release in its state until the zone is left. CONTINUE: If the release is revoked or expires after the vehicle has already entered the zone, the vehicle continues its path, keeping the zoneRequest object granting the zone release in its state. If the order ends inside the zone, the vehicle waits for a new order.| 
```

> | | releaseLossBehavior | string | Enum {'STOP', 'CONTINUE', 'EVACUATE'}</br>When the access to this zone is revoked or expired, the vehicle can either STOP, CONTINUE, or EVACUATE the zone. This action is only executed, when the vehicle is already in the zone and the release expires or is revoked. If not defined, the vehicle is expected to STOP and report an error. STOP: Vehicle stops and sends a CRITICAL error. EVACUATE: Execute the evacuation behavior of the vehicle to leave the zone, keeping the release in its state until the zone is left. CONTINUE: If the revoke/expiry happens, after the vehicle entered the zone, the vehicle continues its path keeping the release of the zone in its state. If the order ends inside the zone, the vehicle waits for a new order. | 

Reviewer: CONTINUE: If release is revoked or expires happens, after the vehicle had already entered the zone, the vehicle continues its path keeping the `zoneRequest` object granting release to the zone in its state. If the order ends inside the zone, the vehicle waits for a new order. |

| COORDINATED_REPLANNING | none | | No autonomous replanning is allowed within this zone. Vehicles are only allowed adjusting their path if granted permission by master control. | 
| SPEED_LIMIT | | | Vehicles must not drive faster than the defined maximum speed within this zone. | 
| | maxSpeed | float64 | Maximum permitted speed for vehicles within the zone in m/s. The speed limit shall already be reached upon entering the zone.|
| ACTION | | | Vehicles shall perform predefined actions when entering, traversing, or exiting the zone. The factsheet defines which actions can be executed when. |

> | | entryActions[action] | array | Actions to be triggered when entering the zone. Empty Array, if no Actions required. | 

Reviewer: | | entryActions[action] | array | Actions to be triggered when entering the zone. Empty array, if no actions required. |

Author: ```suggestion
| | entryActions[action] | array | Actions to be triggered when entering the zone. Empty array, if no actions required. |
```

> | | duringActions[action] | array | Actions to be executed while crossing the zone. Empty Array, if no Actions required. | 

Reviewer: | | duringActions[action] | array | Actions to be executed while crossing the zone. Empty array, if no actions required. |

Author: ```suggestion
| | duringActions[action] | array | Actions to be executed while crossing the zone. Empty array, if no actions required. |
```

> | | exitActions[action] | array | Actions to be triggered when leaving the zone. Empty Array, if no Actions required. | 

Author: ```suggestion
| | exitActions[action] | array | Actions to be triggered when leaving the zone. Empty array, if no actions required. |
```

> #### Kinematic center-based zones

Reviewer: | | exitActions[action] | array | Actions to be triggered when leaving the zone. Empty array, if no actions required. |

> In kinematic center-based zones, the vehicle's kinematic center decides the entry and exit of the zones. When the vehicle's kinematic center is within a zone, the vehicle shall follow the defined behavior. 

Reviewer: In kinematic center-based zones, the vehicle's kinematic center determines its entry and exit of the zones. When the vehicle's kinematic center is inside a zone, the vehicle shall follow the defined behavior.

Author: ```suggestion
In kinematic center-based zones, the vehicle's kinematic center determines its entry and exit of the zones. When the vehicle's kinematic center is inside a zone, the vehicle shall follow the defined behavior.
```

'PRIORITY' and 'PENALTY' zones are zones which only influence the path planning of mobile robots.

> 'DIRECTED' zones define a preferred direction of travel within the zone. 'BIDIRECTED' zones define a travel direction and its opposite direction to be used. Other directions shall be avoided. Additional parameters indicate the limits within which the vehicle may deviate from its direction of travel. The direction of travel is the speed vector in the project-specific coordinate system.

Reviewer: DIRECTED' zones define a preferred direction of travel within the zone. 'BIDIRECTED' zones define a travel direction and its opposite direction to be used. Other directions shall be avoided. The `directedLimitation` enum indicates the limits within which the vehicle may deviate from its direction of travel. The direction of travel is the speed vector in the project-specific coordinate system.

Author: ```suggestion
'DIRECTED' zones define a preferred direction of travel within the zone. 'BIDIRECTED' zones define a travel direction and its opposite direction to be used. Other directions shall be avoided. The `directedLimitation` and `bidirectedLimitation` enums specify the limits within which the vehicle may deviate from its direction of travel. The direction of travel is the velocity vector in the project-specific coordinate system.
```

![Figure x2 Depiction of a vehicle entering a zone based on its kinematic center (left) and a loaded vehicle exiting a zone based on its kinematic center (right)](./assets/kinematic_center_entry.png)
>Figure x2 Depiction of a vehicle entering a zone based on its kinematic center (left) and a loaded vehicle exiting a zone based on its kinematic center (right)

| **Zone Type**| **Zone Parameters** | **Data type** | **Description** | 
| --- | --- | --- | --- |

> | PRIORITY | | | Navigating through this zone is associated with an incentivisation for the vehicle's route compared to an otherwise equivalent area with no such zone on the map.| 

Reviewer: | PRIORITY | | | The workspace encompassed by this zone is associated with an incentivisation for the vehicle to plan its route  through this zone compared to an otherwise equivalent area with no such zone on the map.|

Author: ```suggestion
| PRIORITY | | | The workspace encompassed by this zone is associated with an incentive for the vehicle to plan its route through this zone compared to an otherwise equivalent area without such a zone on the map.| 
```

> | | priorityFactor | float64 | [0.0...1.0]<br> Relative factor, determining the zone's preference over an area with no zone. 0.0 means no preference, as if there was no zone, 1.0 is maximum preference. |

Reviewer: | | priorityFactor | float64 | [0.0...1.0]<br> Relative factor, determining the zone's preference over an workspace with no zone. 0.0 means no preference, as if there was no zone, 1.0 is maximum preference. |

Author: ```suggestion
| | priorityFactor | float64 | [0.0...1.0]<br>Relative factor that determines the preference of the zone over a workspace without a zone. 0.0 means no preference, as if there was no zone, 1.0 is maximum preference.|
```

> | PENALTY | | | Navigating through this zone is associated with an increased cost for the vehicle's route compared to an otherwise equivalent area with no such zone on the map. This zone is intended to decentivise use by vehicles.| 

Reviewer: | PENALTY | | | The workspace encompassed by this zone is associated with a disincentivization for the vehicle to plan its route  through this zone compared to an otherwise equivalent area with no such zone on the map.|

Author: ```suggestion
| PENALTY | | | The workspace encompassed by this zone is associated with a disincentive for the vehicle to plan its route through this zone compared to an otherwise equivalent area without such a zone on the map.|
```

> | | penaltyFactor | float64 | [0.0...1.0]<br> Relative factor, determining the zone's penalty over an area with no zone. 0.0 means no penalty, as if there was no zone, 1.0 is the maximum penalty, causing the mobile robot to take this path only if no other path is possible.  |

Reviewer: | | penaltyFactor | float64 | [0.0...1.0]<br> Relative factor, determining the zone's penalty over a workspace with no zone. 0.0 means no penalty, as if there was no zone, 1.0 is the maximum penalty, causing the mobile robot to take this path only if cannot find any other feasible route. |

Author: ```suggestion
| | penaltyFactor | float64 | [0.0...1.0]<br> Relative factor that determines the penalty of the zone compared to a workspace without that zone. 0.0 means no penalty, as if there was no zone, 1.0 is the maximum penalty, causing the mobile robot to take this path only if it cannot find any other feasible route. |
```

| DIRECTED | | | Vehicles shall traverse this zone in a specific direction of travel. | 

> | | direction | float64 | Preferred direction of travel within the zone in radians. The direction of travel is the speed vector in the project-specific coordinate system. | 

Reviewer: | | direction | float64 | Preferred direction of travel within the zone in radians. The direction of travel is the velocity vector in the project-specific coordinate system. |

Author: Above, we are talking about the speed vector, changed there as well.

Author: ```suggestion
| | direction | float64 | Preferred direction of travel within the zone in radians. The direction of travel is the velocity vector in the project-specific coordinate system. | 
```

> | | directedLimitation | string | Enum {'SOFT','RESTRICTED','STRICT'}<\br>SOFT: Mobile robots may deviate from the defined direction of travel, but should avoid it, RESTRICTED: The mobile robot may deviate from the defined direction of travel, e.g., for obstacle avoidance, but shall never traverse opposite to the defined direction of travel, STRICT: The mobile robot shall maintain the defined direction of travel as precise as possible. |

Reviewer: Erstmal 5/5 fürs neue Wording von dem Enum!

> | | directedLimitation | string | Enum {'SOFT','RESTRICTED','STRICT'}<\br>SOFT: Mobile robots may deviate from the defined direction of travel, but should avoid it, RESTRICTED: The mobile robot may deviate from the defined direction of travel, e.g., for obstacle avoidance, but shall never traverse opposite to the defined direction of travel, STRICT: The mobile robot shall maintain the defined direction of travel as precise as possible. |

Reviewer: | | directedLimitation | string | Enum {'SOFT','RESTRICTED','STRICT'}<\br>SOFT: Mobile robots may deviate from the defined direction of travel but should avoid it, RESTRICTED: The mobile robot may deviate from the defined direction of travel, e.g., for obstacle avoidance but shall never traverse opposite to the defined direction of travel, STRICT: The mobile robot shall maintain the defined direction of travel as precise as its technical capabilities allow. |

Author: ```suggestion
| | directedLimitation | string | Enum {'SOFT','RESTRICTED','STRICT'}<\br>SOFT: Mobile robots may deviate from the defined direction of travel, but should avoid it, RESTRICTED: The mobile robot may deviate from the defined direction of travel, e.g., to avoid an obstacle, but shall never traverse opposite to the defined direction of travel, STRICT: The mobile robot shall maintain the defined direction of travel as precisely as its technical capabilities allow. |
```

| BIDIRECTED | | | While in this zone, vehicles shall only move in the defined direction of travel and its direct opposite (+ Pi), vehicles should not cross this zone in any other direction. | 

> | | direction | float64 | Preferred direction of travel within the zone in radians. The direction of travel is the speed vector in the project-specific coordinate system. | 

Reviewer: | | direction | float64 | Preferred direction of travel within the zone in radians. The direction of travel is the velocity vector in the project-specific coordinate system. |

Author: ```suggestion
| | direction | float64 | Preferred direction of travel within the zone in radians. The direction of travel is the velocity vector in the project-specific coordinate system. | 
```

| | bidirectedLimitation | string | Enum {'SOFT', 'RESTRICTED'}<\br>SOFT: Mobile robots may deviate from the defined directions of travel, but should avoid it, RESTRICTED: The mobile robot should not traverse in any other direction than the directions of travel, except for obstacle avoidance.|
 
### 6.x. Implementation of the zone set object for zone set transfer

> In contrast to the maps, the structure of the zones is predefined. This structure shall be maintained when providing the download as well as sending the zones via the separate topic `zoneSet`. It is not part of the state message of the vehicles.

Reviewer: While this guideline explicitly defines no map format, a zone as well as a zone set is a well-defined JSON object. This structure shall be maintained when providing the download as well as sending the zones via the separate topic `zoneSet`. It is not part of the state message of the vehicles.

Author: ```suggestion
While this guideline  does not explicitly define a map format, both a zone and a zone set are well-defined JSON objects. This structure shall be maintained when providing the download as well as sending the zones via the separate `zoneSet` topic. This structure is not part of the state message.
```

Zone sets shall only be changed and distributed by master control to keep consistency in the system.

A `zoneSet` is an array of `zone` objects with a globally unique identifier, `zoneSetId`. It must be associated with a single map referenced through the `mapId`. The `mapVersion` shall not be referenced, as the same zone set might be intended to be used for several versions of one map. In general, several zone sets can be defined in addition to a single map and it is upon master control to ensure that the right zone set is enabled for each map on the vehicle. As with maps, the `zoneSetStatus` indicates which zone set is currently used by the vehicle. Only a single zone set can be active at once for each `mapId` on the vehicle. Zones shall not extend beyond the spatial boundaries of a map.

> The content of a zone set with a unique zoneSetId shall not change. If changes are required within a zone set, it shall be identified by a new zoneSetId.

Reviewer: The content of a zone set with a unique `zoneSetId` shall not change. If changes are required within a zone set, it shall be referenced with a new `zoneSetId`.

Author: ```suggestion
The content of a zone set with a unique `zoneSetId` shall not change. If changes are required within a zone set, it shall be referenced with a new `zoneSetId`.
```

| **Object structure** | **Data type** | **Description** |
| --- | --- | --- |
| zoneSet{ | JSON object | Zone set detailing a dedicated map. | 
| mapId | string | Globally unique identifier of the map the zone set particularizes. | 
| zoneSetId | string | Globally unique identifier of the zone set. |
| *zoneSetDescription* | string | Human readable description of the zone set. | 
| zones[zone] <br> } | array | Array of zone objects. | 

A single zone object has the following structure: 

| **Object structure** | **Data type** | **Description** |
| --------------------- | ------------- | ------------------- |
| zone{ | JSON object | |
| zoneId | string | Locally (within the zone set) unique identifier. |

> | zoneType | enum | Enum {'BLOCKED', 'LINE_GUIDED', 'RELEASE', 'COORDINATED_REPLANNING', 'SPEED_LIMIT', 'ACTION', 'PRIORITY', 'PENALTY', 'DIRECTED', 'BIDIRECTED'}, Zone type according to Section 6.x.1 Zone types. |

Reviewer 2: ```suggestion
| zoneType | string | Enum {'BLOCKED', 'LINE_GUIDED', 'RELEASE', 'COORDINATED_REPLANNING', 'SPEED_LIMIT', 'ACTION', 'PRIORITY', 'PENALTY', 'DIRECTED', 'BIDIRECTED'}, Zone type according to Section 6.x.1 Zone types. |
```

| *zoneDescription* | string | User-defined human-readable name or descriptor. | 
| **vertices[vertex]**| array | Array of vertices that define the geometric shape of the zone in a clockwise direction. |

> | maxSpeed | float64 | Required only for SPEED_LIMIT zone as defined in chapter 6.x.1.| 

Reviewer: Liest sich zwar ein wenig holprig mit den Parametern, aber finde ich gut gelöst

| **entryActions[Action]**| array | Required only for ACTION zone as defined in chapter 6.x.1.| 
| **duringActions[Action]** | array | Required only for ACTION zone as defined in chapter 6.x.1.| 
| **exitActions[Action]** | array | Required only for ACTION zone as defined in chapter 6.x.1.| 

> | *releaseLossBehavior* | enum | Optional only for RELEASE zone as defined in chapter 6.x.1.|

Reviewer 2: ```suggestion
| *releaseLossBehavior* | string | Optional only for RELEASE zone as defined in chapter 6.x.1.|
```

| **Object structure** | **Data type** | **Description** |
| --------------------- | ------------- | ------------------- |
| vertex{| JSON-object| |
| x | float64 | X-coordinate described in the project-specific coordinate system |
| y <br>} | float64 | Y-coordinate described in the project-specific coordinate system |

## 6.x.3 Interactive zones - Communication 

For communicating requests for the interactive zones 'RELEASE' and 'COORDINATED_REPLANNING', the field `zoneRequests` in the state message is used. The separate topic `response` is used by master control to respond to these requests.

Before entering an interactive zone, the vehicle shall state a request.

> A request before entry of an interactive zone is necessary, even if the order contains released nodes within the zone.

Reviewer: Sehr guter Hinweis

The vehicle decides at which point before entering the zone to make its requests.
If the response is not received in time, the vehicle shall not enter the zone.

Requests shall only be made for zones of enabled zone sets. Zone requests can also be made for zone sets belonging to maps that the vehicle is not currently on.

> The `requestId` allows master control to distinguish between different requests and allows the vehicle to request multiple alternatives at the same time.

Reviewer: The `requestId` allows master control to distinguish between different requests and allows the vehicle to request multiple alternative request concerning the same zone simultaneously.

Author: ```suggestion
The `requestId` allows master control to distinguish between different requests and allows the vehicle to request several alternative requests for the same zone at the same time.
```

> Each request attempt shall use an unique identifier per mobile robot. Ids can be reused after a vehicle restart.

Reviewer: Each request attempt shall use a unique identifier per mobile robot. Ids can be reused after a vehicle restart.

Author: ```suggestion
Each request attempt shall use a unique identifier per mobile robot. Ids can be reused after a vehicle restart.
```

For requests to enter a 'RELEASE' zone, a `zoneRequest` object of `requestType` 'ACCESS' shall be added to the state message.
For permission to enter a 'COORINATED_REPLANNING' zone with a planned path or for replanning its path within the zone, the `requestType` shall be set to 'REPLANNING'.

> For a 'REPLANNING' request, the planned path shall be added as NURBS to the `trajectory` field of the `zoneRequest`. Multiple requests with different trajectories for the same zone can be made.

Reviewer: For a 'REPLANNING' request, the planned path shall be added as NURBS to the `trajectory` field of the `zoneRequest`. Multiple requests with different trajectories for the same zone can be made. Each path must be requested with its own zoneRequest object.

Author: ```suggestion
For a 'REPLANNING' request, the planned path shall be added as NURBS to the `trajectory` field of the `zoneRequest`. Multiple requests with different trajectories for the same zone can be made. Each path shall be requested with its own `zoneRequest` object.
```

> 

Reviewer: If a noble robot requires access to a workspace covered by two or more 'RELEASE' zones, it must request access to each zone individually before being allowed entering the area. 

Analogously, if a mobile robot wishes to navigate through a workspace on the map that is covered by two or more 'COORDINATED REPLANNING' zones, it must have the route it wishes to take within this area have approved by master control inidivudally in a separate request per zone.

Author: ```suggestion
If a mobile robot requires access to a workspace covered by two or more 'RELEASE' zones, it shall request access and receive approval before entering the area.

If a mobile robot navigates through a workspace on the map that is covered by two or more 'COORDINATED REPLANNING' zones, it shall request its path within this area individually for each zone and receive approval from the master control before entering or changing paths.
```

When receiving a response with `responseType` 'REVOKED', the vehicle shall set the `requestStatus` to 'REVOKED' and not enter the 'RELEASE' zone. When the `leaseExpiry` has passed, the requestStatus shall be set to 'EXPIRED' and the zone shall not be entered. If the vehicle is already inside the 'RELEASE' zone when the `leaseExpiry` has passed or the request is 'REVOKED', it shall report a warning and react according to the `releaseLossBehavior` defined in the zone definition.

![Figure x3 Zone request behavior for a RELEASE zone.](./assets/request_release_zone_access.png)
>Figure x3 Zone request behavior for a RELEASE zone.

The interaction between the mobile robot and the master control for 'COORDINATED_REPLANNING' zones shall be according to figure x4.

The vehicle shall choose one of the trajectories of all 'GRANTED' requests to the zone and set the corresponding `requestStatus`to 'GRANTED' while removing all other requests from its state.

> When receiving a response with `responseType` 'REVOKED', the vehicle shall set the `requestStatus` to 'REVOKED' and not enter the 'COORDINATED_REPLANNING' zone.  When the `leaseExpiry` has passed, the requestStatus shall be set to 'EXPIRED' and the zone shall not be entered. If the vehicle is already inside the 'RELEASE' zone when the `leaseExpiry` has passed or the request is 'REVOKED', it shall stop its movement and report a warning. To continue, the vehicle has to state a new request.

Reviewer: When receiving a response with `responseType` 'REVOKED', the vehicle shall set the `requestStatus` to 'REVOKED' and not enter the 'COORDINATED_REPLANNING' zone. When the `leaseExpiry` has passed, the `requestStatus` shall be set to 'EXPIRED' and the zone shall not be entered. If the vehicle is already inside the 'RELEASE' zone when the `leaseExpiry` has passed or the request is 'REVOKED', it shall stop its movement and report a warning. To continue, the vehicle has to state a new request.

Author: ```suggestion
When receiving a response with `responseType` 'REVOKED', the vehicle shall set the `requestStatus` to 'REVOKED' and not enter the 'COORDINATED_REPLANNING' zone. When the `leaseExpiry` has passed, the `requestStatus` shall be set to 'EXPIRED' and the zone shall not be entered. If the vehicle is already inside the 'RELEASE' zone when the `leaseExpiry` has passed or the request is 'REVOKED', it shall stop its movement and report a warning. To continue, the vehicle shall state a new request.
```

![Figure x4 Zone request behavior for a COORDINATED_REPLANNING zone.](./assets/request_coordinated_replanning_zone_replanning.png)
>Figure x4 Zone request behavior for a COORDINATED_REPLANNING zone.

### Response message definitions

Object structure/Identifier | Data type | Description
| --- | --- | --- |
|headerId | | uint32 | Header ID of the message.<br> The headerId is defined per topic and incremented by 1 with each sent (but not necessarily received) message.
|timestamp | | string | Timestamp (ISO 8601, UTC); YYYY-MM-DDTHH:mm:ss.fffZ (e.g., "2017-04-15T11:40:03.123Z").
|version | | string | Version of the protocol [Major].[Minor].[Patch] (e.g., 1.3.2).
|manufacturer | | string | Manufacturer of the AGV.
|serialNumber | | string | Serial number of the AGV.
|responses[response] | array | Array of response objects. |

Object structure/Identifier | Data type | Description
| --- | --- | --- |
| response <br> { | JSON object | Object which contains the master control answer to a specific request. |
| requestId | string | Unique per vehicle identifier within all active requests. |

> | grantType | string | Enum {'GRANTED','QUEUED','REVOKED','REJECTED'}<\br>'GRANTED': master control grants request. 'REVOKED': master control revokes previously granted request. 'REJECTED': master control rejects a request. 'QUEUED': Acknowledge the vehicles request to the master control, but no permission is given yet. Request was added to some sort of a queue. |

Reviewer: | grantType | string | Enum {'GRANTED','QUEUED','REVOKED','REJECTED'}<\br>'GRANTED': master control grants request. 'REVOKED': master control revokes previously granted request. 'REJECTED': master control rejects a request. 'QUEUED': master control acknowledges the vehicles request but no permission is given yet. |

Author: ```suggestion
| grantType | string | Enum {'GRANTED','REVOKED','REJECTED','QUEUED'}<\br>'GRANTED': Master control grants request. 'REVOKED': Master control revokes previously granted request. 'REJECTED': Master control rejects a request. 'QUEUED': Master control acknowledges the vehicles request, but no permission is given yet.|
```

| |**BLOCKED**|**RELEASE**|**LINE_GUIDED**|**COORDINATED_REPLANNING**|**SPEED_LIMIT**|**ACTION**|**PRIORITY**|**PENALTY**|**DIRECTED**|**BIDIRECTED**|**No Zone**|**EDGE-PROPERTIES**
---|---|---|---|---|---|---|---|---|---|---|---|---
**BLOCKED**|BLOCKED|BLOCKED|BLOCKED|BLOCKED|BLOCKED|BLOCKED|BLOCKED|BLOCKED|BLOCKED|BLOCKED|BLOCKED|BLOCKED|
**RELEASE**||No Conflict|No Conflict|No Conflict|No Conflict|No Conflict|No Conflict|No Conflict|No Conflict|No Conflict|No Conflict|No Conflict
**LINE_GUIDED**|||No conflict|LINE_GUIDED|No Conflict| (1) |LINE_GUIDED|LINE_GUIDED|LINE_GUIDED|No conflict|LINE_GUIDED|No conflict
**COORDINATED_REPLANNING**||||(2)|No conflict|(1)|No conflict|No conflict|No conflict|No conflict|COORDINATED_REPLANNING|(3)
**SPEED_LIMIT** |||||(4)|No conflict|No conflict|No conflict|No conflict|No conflict|SPEED_LIMIT|(4)
**ACTION** ||||||(5)|No conflict|No conflict|No conflict|No conflict|ACTION|(5)
**PRIORITY** |||||||(6)|(6)|No conflict|No conflict|(7)|No conflict
**PENALTY** ||||||||(6)|No conflict|No conflict|(7)|No conflict
**DIRECTED** |||||||||(8)|(8)|(7)|(9)
**BIDIRECTED** ||||||||||(8)|(7)|(9)

> 1) If actions would conflict with other zones behavior, report a CRITICAL error (order error) and stop the vehicle.

Author: ```suggestion
1) If actions would conflict with other zones behavior, report a 'CRITICAL' error (order error) and stop the vehicle.
```

#### Contour-based zones

For contour-based zones, the vehicles contour (including its load) decides the entry and exit of the zones. These zones shall not be entered by any part of the contour, if the master control does not allow it, e.g., a 'BLOCKED' zone may never be entered, a 'RELEASE' zone may only be entered, if it is released by the master control, etc. or they lead to a defined behavior when entered, e.g., in the ACTION zone when entering all entryActions shall be executed and so on.

![Figure x1 Depiction of a vehicle entering a zone based on its contour (left) and a loaded vehicle with corresponding extended bounding box exiting a zone (right)](./assets/contour_entry.png)
>Figure x1 Depiction of a vehicle entering a zone based on its contour (left) and a loaded vehicle with corresponding extended bounding box exiting a zone (right)

The following contour-based zones are defined:

| **Zone Type**| **Zone Parameters** | **Data type** | **Description** | 
| --- | --- | --- | --- |

> | BLOCKED | none | - | Vehicles must not enter this zone. | 

Reviewer 3: For 2.1, we replaced "must" etc. with "shall" -> principle not fully applied in zone additions

| LINE_GUIDED | none | - | No autonomous navigation is allowed in this zone. Vehicles are only allowed driving in a (virtually) line-guided way. Vehicles may only enter this zone if the route is explicitly specified by the master control in the form of a node-edge graph. | 
| RELEASE | none | - | Vehicles are only allowed entering this zone once they have been granted access through master control. | 

> | | releaseLossBehavior | enum | When the access to a releaseZone is revoked or expired, the vehicle can either STOP, CONTINUE, or EVACUATE the zone. This Action is only executed, when the vehicle is already in the zone and the release expires or is revoked. If not defined, the vehicle is expected to STOP and report an error. STOP: Vehicle stops, sends an fatal error. EVACUATE: Execute the evacuation behavior of the vehicle to leave the zone, keeping the release in its state until the zone is left. CONTINUE: If the revoke/expiry happens, after the vehicle entered the zone, the vehicle continue its path keeping the reservation of the vehicle in its state. If the order ends inside the zone, the vehicle waits for a new order. | 

Reviewer 3: ```suggestion
| | releaseLossBehavior | enum | When access to this zone is revoked or expired, the vehicle can either STOP, CONTINUE, or EVACUATE the zone. This Action is only executed, when the vehicle is already in the zone and the release expires or is revoked. If not defined, the vehicle is expected to STOP and report an error. STOP: Vehicle stops, sends an fatal error. EVACUATE: Execute the evacuation behavior of the vehicle to leave the zone, keeping the release in its state until the zone is left. CONTINUE: If the revoke/expiry happens, after the vehicle entered the zone, the vehicle continues its path keeping the reservation of the vehicle in its state. If the order ends inside the zone, the vehicle waits for a new order. | 
```

| COORDINATED_REPLANNING | - | - | No autonomous replanning is allowed within this zone. Vehicles are only allowed adjusting their global or local path if granted by master control. | 
| SPEED_LIMIT | | - | Vehicles must not drive faster than the defined maximum speed within this zone. | 

> | | maxSpeed | float64 | Maximum feasible speed for vehicles within the zone in m/s. The speed limit shall be reached when entering the zone.|

Reviewer 3: ```suggestion
| | maxSpeed | float64 | Maximum permitted speed for vehicles within the zone in m/s. The speed limit shall be reached when entering the zone.|
```

> | ACTION | | - | Vehicles shall perform predefined actions when entering, traversing, or exiting the zone. The factsheet defines, which actions can be executed when. |

Reviewer 3: ```suggestion
| ACTION | | - | Vehicles shall perform predefined actions when entering, traversing, or exiting the zone. The factsheet defines which actions can be executed when. |
```

| | entryActions[action] | array | Actions to be triggered when entering the zone. Empty Array, if no Actions required. | 
| | duringActions[action] | array | Actions to be executed while crossing the zone. Empty Array, if no Actions required. | 
| | exitActions[action] | array | Actions to be triggered when leaving the zone. Empty Array, if no Actions required. | 

#### Kinematic center-based zones

> In kinematic center based-zones, the vehicle's kinematic center decides the entry and exit of the zones. When within a zone, the vehicle shall follow the defined behavior. 

Reviewer 3: ```suggestion
In kinematic center-based zones, the vehicle's kinematic center decides the entry and exit of the zones. When the vehicle's kinematic center is within a zone, the vehicle shall follow the defined behavior. 
```

> 'PRIORITY' and 'PENALTY' zones represent zones, that are used for the influencing the movement of the robots by incetivising or penalizing certain areas of the map. They have no effect on the behavior of the robots when passing through.

Reviewer 3: ```suggestion
'PRIORITY' and 'PENALTY' zones represent zones that are used for the influencing the movement of the robots by incentivising or penalizing certain areas of the map. They have no effect on the behavior of the robots when passing through.
```

| **Zone Type**| **Zone Parameters** | **Data type** | **Description** | 
| --- | --- | --- | --- |
| PRIORITY | | | Navigating through this zone is associated with a reduced cost for the vehicle's route compared to any other identical area on the map. This zone is intended to incentivise use by vehicles.| 
| | priorityFactor | float64 | [0.0...1.0]<br> Relative factor, determining the zone's preference over an area with no zone. 0.0 means no preference, 1.0 is maximum preference. |
| PENALTY | | | Navigating through this zone is associated with an increased cost for the vehicle's route compared to any other identical area on the map. This zone is intended to decentivise use by vehicles.| 
| | penaltyFactor | float64 | [0.0...1.0]<br> Relative factor, determining the zone's penalty over an area with no zone. 0.0 means no penalty, 1.0 is maximum penalty. | 
| DIRECTED | | - | Vehicles shall (preferably) traverse this zone in specific directions of travel. | 
| | direction | float64 | Preferred direction of travel within the zone in radians. The direction of travel is the speed vector in the project-specific coordinate system. | 

> | | limitation | enum | ['NONE', 'OBSTACLE','STRICT']<\br>NONE: Vehicles may traverse against the direction of travel within the zone in the opposite direction if necessary, but shall avoid it (e.g. make sharp turns into islands or if otherwise completely blocked), OBSTACLE: Vehicles may not traverse against the direction of travel except to avoid obstacles, STRICT: Vehicles shall not traverse against the direction of travel within the zone. |

Reviewer 3: ```suggestion
| | limitation | enum | ['NONE', 'OBSTACLE','STRICT']<\br>NONE: Vehicles may traverse against the direction of travel within the zone in the opposite direction if necessary, but should avoid it (e.g. make sharp turns into islands or if otherwise completely blocked), OBSTACLE: Vehicles may not traverse against the direction of travel except to avoid obstacles, STRICT: Vehicles shall not traverse against the direction of travel within the zone. |
```

| BIDIRECTED | | | While in this zone, vehicles shall only move in the direction of travel and the opposite, vehicles shall not cross this zone in any other directions. | 
| | direction | float64 | Allowed direction of travel within the zone in radians, the second allowed direction of travel is +/- Pi. The direction of travel is the speed vector in the project-specific coordinate system. | 

> | | limitation | enum | ['NONE', 'OBSTACLE']<\br>NONE: Vehicles may traverse in any direction other then the main driving directions, but shall avoid it (e.g. make sharp turns into islands or if otherwise completely blocked), OBSTACLE: Vehicles may not traverse in any direction other then the main driving directions except to avoid obstacles.|

Reviewer 3: ```suggestion
| | limitation | enum | ['NONE', 'OBSTACLE']<\br>NONE: Vehicles may traverse in any direction other then the main driving directions, but should avoid it (e.g. make sharp turns into islands or if otherwise completely blocked), OBSTACLE: Vehicles may not traverse in any direction other then the main driving directions except to avoid obstacles.|
```

## 6.x.2 Sharing of planned path for freely navigating mobile robots

Vehicles shall communicate their planned trajectory to the master control system. This is done via the state message. For a higher frequency of sharing, the visualisation topic can be used. For this purpose the parameters `plannedPath` and `intermediatePath` are introduced, to be used only for trajectories planned by the mobile robot. The trajectory fields in the edgeState should only be used to 'acknowledge' trajectories that have already been defined a priori within a layout and not during operations.

> This recommendation distinguishes between the `plannedPath`, which represents a longer path within the robot's currently active order that it is confident to share, and the `intermediatePath`, which represents the estimated time of arrival at closer waypoints that the vehicle is able to perceive with its sensors. Both paths start from the current position of the mobile robot (independent of any node that is part of the mission the robot is currently executing) and go as far as the robot has what it considers to be reliable data, as it may also be situation dependent.

Reviewer 3: ```suggestion
This recommendation distinguishes between the `plannedPath`, which represents a longer path within the robot's currently active order that it is confident to share, and the `intermediatePath`, which represents the estimated time of arrival at closer waypoints that the vehicle is able to perceive with its sensors. Both paths start from the current position of the mobile robot (independent of any node that is part of the mission the robot is currently executing) and go as far as the robot has what it considers to be reliable data.
```

> The `plannedPath` is defined as a NURBS identical to the one defined in the `trajectory` field of the `edgeState`. In addition, the robot may contain an array of nodes (referenced by their `nodeId`) that are included in the current plan. 

Reviewer 3: ```suggestion
The `plannedPath` is defined as a NURBS identical to the one defined in the `trajectory` field of the `edgeState`. In addition, the planned path may contain an array of nodes (referenced by their `nodeId`) that will be traversed as part of the current plan. 
```

> The `intermediatePath` is defined as a polyline with an estimated time of arrival defined by a `timestamp` and optionally the orientation of the mobile robot added to each shared waypoint. It is defined as an array of waypoints where the line segments of a `polyline` are linear. The exact length of the shared path is limited by the perception of its intermediate environment by its sensors.

Reviewer 3: ```suggestion
The `intermediatePath` is defined as a polyline with an estimated time of arrival defined by a `timestamp` and optionally the orientation of the mobile robot added to each shared waypoint. It is defined as an array of waypoints where the line segments of a `polyline` are linear. The exact length of the shared path is limited by the robot's perception of its intermediate environment by its sensors.
```

### 6.x. Implementation of the zone set object for zone set transfer

> In contrast to the maps, the structure of the zones is predefined. This structure shall be maintained when downloading as well as sending the zones via the separate topic `zoneSet`. This structure is not part of the state message of the vehicles.

Reviewer 3: ```suggestion
In contrast to the maps, the structure of the zones is predefined. This structure shall be maintained when providing the download as well as sending the zones via the separate topic `zoneSet`. It is not part of the state message of the vehicles.
```

A zoneSet shall only be changed and distributed by master control to keep consistency in the system.

> A `zoneSet` is an array of `zone` objects with a globally unique identifier, `zoneSetId`. It must be associated with a single map referenced through the `mapId`. The `mapVersion` shall not be referenced, as the same zone set might be intended to be used for several versions of one map. In general, several zone sets can be defined in addition to a single map and it is upon master control to ensure that the right zone set is enabled for each map on the vehicle. As with maps, the `zoneSetStatus` indicates which zone set is currently used by the device. Only a single zone set can be activated at a time for each `mapId` on the vehicle. Zones shall not extend beyond the spatial boundaries of a map.

Reviewer 3: ```suggestion
A `zoneSet` is an array of `zone` objects with a globally unique identifier, `zoneSetId`. It must be associated with a single map referenced through the `mapId`. The `mapVersion` shall not be referenced, as the same zone set might be intended to be used for several versions of one map. In general, several zone sets can be defined in addition to a single map and it is upon master control to ensure that the right zone set is enabled for each map on the vehicle. As with maps, the `zoneSetStatus` indicates which zone sets are currently used by the device. Only a single zone set can be activated at a time for each `mapId` on the vehicle. Zones shall not extend beyond the spatial boundaries of a map.
```

| zones[zone] <br> } | array | Array of zone objects. | 

A single zone object has the following structure: 

| **Object structure** | **Data type** | **Description** |
| --------------------- | ------------- | ------------------- |
| zone{ | JSON object | |
| zoneId | string | Locally (within the zone set) unique identifier. |
| zoneType | string | Enum {'BLOCKED', 'LINE_GUIDED', 'RELEASE', 'COORDINATED_REPLANNING', 'SPEED_LIMIT', 'ACTION', 'PRIORITY', 'PENALTY', 'DIRECTED', 'BIDIRECTED'}, Zone type according to Section 6.x.1 Zone types. |
| *zoneDescription* | string | User-defined human-readable name or descriptor. | 
| *maxSpeed* | float64 | Required in SPEED_LIMIT zone.| 
| ***entryActions[Action]*** | array | Array of actions to be executed when entering the zone. Empty array, if no actions required.| 
| ***duringActions[Action]*** | array | Actions to be executed while crossing the zone. Empty array, if no actions required.| 
| ***exitActions[Action]*** | array | Actions to be triggered when leaving the zone. Empty array, if no actions required.| 
| **vertices[vertex]** | array | Array of vertices (in x-y-coordinates) defining the geometrical shape of the zone. |

> | *releaseLossBehavior* <br> } | enum | When the access to a releaseZone is revoked or expired, the vehicle can either STOP, CONTINUE, or EVACUATE the zone. This Action is only executed, when the vehicle is already in the zone and the release expires or is revoked. If not defined, the vehicle is expected to STOP and report an error. STOP: Vehicle stops, sends an fatal error. EVACUATE: Execute the evacuation behavior of the vehicle to leave the zone, keeping the release in its state until the zone is left. CONTINUE: If the revoke/expiry happens, after the vehicle entered the zone, the vehicle continue its path keeping the reservation of the vehicle in its state. If the order ends inside the zone, the vehicle waits for a new order. |

Reviewer 3: ```suggestion
| *releaseLossBehavior* <br> } | enum | When the access to a release zone is revoked or expired, the vehicle can either STOP, CONTINUE, or EVACUATE the zone. This Action is only executed, when the vehicle is already in the zone and the release expires or is revoked. If not defined, the vehicle is expected to STOP and report an error. STOP: Vehicle stops and sends a fatal error. EVACUATE: Execute the evacuation behavior of the vehicle to leave the zone, keeping the release in its state until the zone is left. CONTINUE: If the revoke/expiry happens after the vehicle entered the zone, the vehicle continues its path keeping the reservation of the vehicle in its state. If the order ends inside the zone, the vehicle waits for a new order. |
```

After leaving the zone the vehicle shall remove the corresponding `zoneRequest` entry from its state message.

![Figure x3 Zone request behavior for a RELEASE zone.](./assets/request_release_zone_access.png)
>Figure x3 Zone request behavior for a RELEASE zone.

Likewise, when a vehicle needs to replan while in a 'COORDINATED_REPLANNING' zone, it adds a zone request object with `requestStatus` set to 'REQUESTED' for the corresponding zone to its state message.
If it receives permission, it reports `requestStatus` to 'GRANTED' and can use the requested trajectory.
If there are multiple requests for the same 'COORDINATED_REPLANNING' zone accepted by the master control, the vehicle shall only choose one of them that it will actually traverse and set it to true in its state and remove the other ones from its state.
The vehicles driving trajectory shall only deviate from the released path by the vehicles precision.

![Figure x4 Zone request behavior for a COORDINATED_REPLANNING zone.](./assets/request_coordinated_replanning_zone_replanning.png)
>Figure x4 Zone request behavior for a COORDINATED_REPLANNING zone.

> The operating mode is influencing the zone release behavior (Section [6.10.6 Implementation of the state message](#6106-Implementation-of-the-state-message)).

Reviewer 3: ```suggestion
The operating mode may influence the zone release behavior (Section [6.10.6 Implementation of the state message](#6106-Implementation-of-the-state-message)).
```

Master control response to zone requests by sending a zone response message via the `response` topic.

> A zone response message contains in addition to the usual header information an array of zone response objects. Each zone response object can only grant a single request at a time referenced by the `requestId`.

Reviewer 3: ```suggestion
The response message contains an array of zone response objects. Each zone response object can only grant a single request at a time referenced by the `requestId`.
```

Additionally, master control has the option to revoke a permission once granted directly by sending a message on the `response` topic referencing the `requestId` with `grantType` set to 'REVOKED'.
When receiving a zone revoke message, the vehicle shall set the `requestStatus` to 'REVOKED'. If the vehicle is already inside the requested zone and receives a zone revoke message, it shall reporting a warning and react according to the `releaseLossBehavior` value defined in the zone definition.
The corresponding `requestId` shall not be reused to renew initial request.
The master control shall assume a revoked request as still granted until receiving an corresponding updated state message from the vehicle.

If the vehicle is already inside the requested zone and the grant lease is expired the vehicle shall report a warning and act accordingly to the defined `releaseLossBehavior`.

When a vehicle within a 'COORDINATED_REPLANNING' zone receives a revoke message or if the grant lease is expired, the vehicle shall stop and state a new request if required.

> ### Grant message definitions

Reviewer 3: ```suggestion
### Response message definitions
```

| |**BLOCKED**|**LINE_GUIDED**|**COORDINATED_REPLANNING**|**SPEED_LIMIT**|**ACTION**|**PRIORITY**|**PENALTY**|**DIRECTED**|**BIDIRECTED**|**No Zone**|**EDGE-PROPERTIES**
---|---|---|---|---|---|---|---|---|---|---|---
**BLOCKED**|BLOCKED|BLOCKED|BLOCKED|BLOCKED|BLOCKED|BLOCKED|BLOCKED|BLOCKED|BLOCKED|BLOCKED
**LINE_GUIDED**||No conflict|LINE_GUIDED|No Conflict| (1) |LINE_GUIDED|LINE_GUIDED|LINE_GUIDED|No conflict|LINE_GUIDED|No conflict
**COORDINATED_REPLANNING**|||(2)|No conflict|(1)|No conflict|No conflict|No conflict|No conflict|COORDINATED_REPLANNING|(3)
**SPEED_LIMIT** ||||(4)|No conflict|No conflict|No conflict|No conflict|No conflict|SPEED_LIMIT|(4)
**ACTION** |||||(5)|No conflict|No conflict|No conflict|No conflict|ACTION|(5)
**PRIORITY** ||||||(6)|(6)|No conflict|No conflict|(7)|No conflict
**PENALTY** |||||||(6)|No conflict|No conflict|(7)|No conflict
**DIRECTED** ||||||||(8)|(8)|(7)|(9)
**BIDIRECTED** |||||||||(8)|(7)|(9)

> 1) If actions from any action would conflict with other zones behavior, report a warning (order error) and stop the vehicle.

Reviewer 3: "actions from any action" does not really sound comprehensible

The master control can request the deletion of a specific map from a vehicle. This is done with the instant action `deleteMap`. When a vehicle runs out of memory, it should report this to the master control, which can then initiate the deletion of maps. The vehicle itself is not allowed to delete maps.
After successfully deleting a map, it is important to remove that map's entry from the vehicle's array of maps in the vehicle state.

> ## 6.2 Sharing of planned path for freely navigating mobile robots (Standalone chapter)

Author: ```suggestion
## 6.y Sharing of planned path for freely navigating mobile robots
```

Author: ```suggestion
## 6.y Sharing of planned path for freely navigating mobile robots
```

| --- | --- | --- | --- |
| PRIORITY | | | Navigating through this zone is associated with an incentivisation for the vehicle's route compared to an otherwise equivalent area with no such zone on the map.| 
| | priorityFactor | float64 | [0.0...1.0]<br> Relative factor, determining the zone's preference over an area with no zone. 0.0 means no preference, as if there was no zone, 1.0 is maximum preference. |
| PENALTY | | | Navigating through this zone is associated with an increased cost for the vehicle's route compared to an otherwise equivalent area with no such zone on the map. This zone is intended to decentivise use by vehicles.| 
| | penaltyFactor | float64 | [0.0...1.0]<br> Relative factor, determining the zone's penalty over an area with no zone. 0.0 means no penalty, as if there was no zone, 1.0 is the maximum penalty, causing the mobile robot to take this path only if no other path is possible.  |
| DIRECTED | | | Vehicles shall traverse this zone in a specific direction of travel. | 
| | direction | float64 | Preferred direction of travel within the zone in radians. The direction of travel is the speed vector in the project-specific coordinate system. | 

>   | | directedLimitation | string | Enum {'SOFT','RESTRICTED','STRICT'}<\br>SOFT: Mobile robots may deviate from the defined direction of travel, but should avoid it, RESTRICTED: The mobile robot may deviate from the defined direction of travel, e.g., for obstacle avoidance, but shall never traverse opposite to the defined direction of travel, STRICT: The mobile robot shall maintain the defined direction of travel as precise as possible. |

Author: ```suggestion
| | directedLimitation | string | Enum {'SOFT','RESTRICTED','STRICT'}<\br>SOFT: Mobile robots may deviate from the defined direction of travel, but should avoid it, RESTRICTED: The mobile robot may deviate from the defined direction of travel, e.g., for obstacle avoidance, but shall never traverse opposite to the defined direction of travel, STRICT: The mobile robot shall maintain the defined direction of travel as precise as possible. |
```

| **Object structure** | **Data type** | **Description** |
| --- | --- | --- |
| zoneSet{ | JSON object | Zone set detailing a dedicated map. | 
| mapId | string | Globally unique identifier of the map the zone set particularizes. | 
| zoneSetId | string | Globally unique identifier of the zone set. |
| *zoneSetDescription* | string | Human readable description of the zone set. | 
| zones[zone] <br> } | array | Array of zone objects. | 

A single zone object has the following structure: 

| **Object structure** | **Data type** | **Description** |
| --------------------- | ------------- | ------------------- |
| zone{ | JSON object | |
| zoneId | string | Locally (within the zone set) unique identifier. |
| zoneType | enum | Enum {'BLOCKED', 'LINE_GUIDED', 'RELEASE', 'COORDINATED_REPLANNING', 'SPEED_LIMIT', 'ACTION', 'PRIORITY', 'PENALTY', 'DIRECTED', 'BIDIRECTED'}, Zone type according to Section 6.x.1 Zone types. |
| *zoneDescription* | string | User-defined human-readable name or descriptor. | 
| **vertices[vertex]** <br> } | array | Array of vertices (in x-y-coordinates) defining the geometrical shape of the zone clockwise. |

> | *maxSpeed* | float64 | Required in SPEED_LIMIT zone as defined in chapter 6.x.1.| 

Author: ```suggestion
| maxSpeed | float64 | Required only for SPEED_LIMIT zone as defined in chapter 6.x.1.| 
```

> | ***entryActions[Action]*** | array | Required in ACTION zone as defined in chapter 6.x.1.| 

Author: ```suggestion
| **entryActions[Action]**| array | Required only for ACTION zone as defined in chapter 6.x.1.| 
```

> | ***duringActions[Action]*** | array | Required in ACTION zone as defined in chapter 6.x.1.| 

Author: ```suggestion
| **duringActions[Action]** | array | Required only for ACTION zone as defined in chapter 6.x.1.| 
```

> | ***exitActions[Action]*** | array | Required in ACTION zone as defined in chapter 6.x.1.| 

Author: ```suggestion
| **exitActions[Action]** | array | Required only for ACTION zone as defined in chapter 6.x.1.| 
```

> | *releaseLossBehavior* | enum | Optional for a RELEASE zone as defined in chapter 6.x.1.|

Author: ```suggestion
| *releaseLossBehavior* | enum | Optional only for RELEASE zone as defined in chapter 6.x.1.|
```

> 

Author: | priorityFactor | float64 | Required only for PRIORITY zone as defined in chapter 6.x.1.|
| penaltyFactor | float64 | Required only for PENALTY zone as defined in chapter 6.x.1.|
| direction | float64 | Required only for DIRECTED and BIDIRECTED zone as defined in chapter 6.x.1.|
| directedLimitation | string | Required only for a DIRECTED zone as defined in chapter 6.x.1.|
| bidirectedLimitation | string | Required only for a BIDIRECTED zone as defined in chapter 6.x.1.|

> 

Author: ```suggestion
| priorityFactor | float64 | Required only for PRIORITY zone as defined in chapter 6.x.1.|
| penaltyFactor | float64 | Required only for PENALTY zone as defined in chapter 6.x.1.|
| direction | float64 | Required only for DIRECTED and BIDIRECTED zone as defined in chapter 6.x.1.|
| directedLimitation | string | Required only for a DIRECTED zone as defined in chapter 6.x.1.|
| bidirectedLimitation | string | Required only for a BIDIRECTED zone as defined in chapter 6.x.1.|

```

The vehicles driving trajectory shall only deviate from the released path by the vehicles precision.

![Figure x4 Zone request behavior for a COORDINATED_REPLANNING zone.](./assets/request_coordinated_replanning_zone_replanning.png)
>Figure x4 Zone request behavior for a COORDINATED_REPLANNING zone.

The operating mode may influence the zone release behavior (Section [6.10.6 Implementation of the state message](#6106-Implementation-of-the-state-message)).

Master control response to zone requests by sending a zone response message via the `response` topic.
The response message contains an array of zone response objects. Each zone response object can only grant a single request at a time referenced by the `requestId`.
A zone request can only be fully granted or rejected, master control cannot conditionally or partially accept a request.

Each response object includes a lease time which specifies until what time a permission is valid. The master control shall resend a release message on the `response` topic before the lease time expires if the permission is still granted.

> Additionally, master control has the option to revoke a permission once granted directly by sending a message on the `response` topic referencing the `requestId` with `grantType` set to 'REVOKED'.

Author: ```suggestion
Master control has the option to revoke a permission once granted directly by sending a message on the `response` topic referencing the `requestId` with `grantType` set to 'REVOKED'.
```

When receiving a zone revoke message, the vehicle shall set the `requestStatus` to 'REVOKED'. If the vehicle is already inside the requested zone and receives a zone revoke message, it shall reporting a warning and react according to the `releaseLossBehavior` value defined in the zone definition.
The corresponding `requestId` shall not be reused to renew initial request.

> The master control shall assume a revoked request as still granted until receiving an corresponding updated state message from the vehicle.

Author: ```suggestion
The master control shall assume a revoked request as still granted until receiving a corresponding updated state message from the vehicle.
```

If the vehicle is already inside the requested zone and the grant lease is expired the vehicle shall report a warning and act accordingly to the defined `releaseLossBehavior`.

When a vehicle within a 'COORDINATED_REPLANNING' zone receives a revoke message or if the grant lease is expired, the vehicle shall stop and state a new request if required.

> 

Author: Master control can set the `requestStatus` to 'QUEUED' to acknowledge the vehicle's request without giving permission, telling the vehicle that its request is being processed.

> 

Author: ```suggestion
Master control can set the `requestStatus` to 'QUEUED' to acknowledge the vehicle's request without giving permission, telling the vehicle that its request is being processed.

```

### Response message definitions

Object structure/Identifier | Data type | Description
| --- | --- | --- |
|headerId | | uint32 | Header ID of the message.<br> The headerId is defined per topic and incremented by 1 with each sent (but not necessarily received) message.
|timestamp | | string | Timestamp (ISO 8601, UTC); YYYY-MM-DDTHH:mm:ss.fffZ (e.g., "2017-04-15T11:40:03.123Z").
|version | | string | Version of the protocol [Major].[Minor].[Patch] (e.g., 1.3.2).
|manufacturer | | string | Manufacturer of the AGV.
|serialNumber | | string | Serial number of the AGV.
|responses[response] | array | Array of response objects. |

Object structure/Identifier | Data type | Description
| --- | --- | --- |
| response <br> { | JSON object | Object which contains the master control answer to a specific request. |
| requestId | string | Unique per vehicle identifier within all active requests. |

> | grantType | enum | Enum {'GRANTED','QUEUED','REVOKED','REJECTED'}<\br>'GRANTED': master control grants request. 'REVOKED': master control revokes previously granted request. 'REJECTED': master control rejects a request. 'QUEUED': Acknowledge the vehicles request to the master control, but no permission is given yet. Request was added to some sort of a queue. |

Author: ```suggestion
| grantType | string | Enum {'GRANTED','QUEUED','REVOKED','REJECTED'}<\br>'GRANTED': master control grants request. 'REVOKED': master control revokes previously granted request. 'REJECTED': master control rejects a request. 'QUEUED': Acknowledge the vehicles request to the master control, but no permission is given yet. Request was added to some sort of a queue. |
```

| directedLimitation | string | Required only for a DIRECTED zone as defined in chapter 6.x.1.|
| bidirectedLimitation | string | Required only for a BIDIRECTED zone as defined in chapter 6.x.1.|

The shape of each zone object is defined through a polygon, which is communicated through its vertices. A zone with less than three vertices is invalid and shall be rejected. If the first entry of the vertex array is not identical to the last, the polygon is implicitly closed by a connecting line to the first vertex. Only simple polygons (i.e. without intersections) shall be used. The array of vertices defining a zone is provided as a list of x-y tuples in the globally defined project-specific coordinate system in a clockwise direction: 

| **Object structure** | **Data type** | **Description** |
| --------------------- | ------------- | ------------------- |
| vertex{| JSON-object| |
| x | float64 | X-coordinate described in the project-specific coordinate system |
| y <br>} | float64 | Y-coordinate described in the project-specific coordinate system |

## 6.x.3 Interactive zones - Communication 

> For communicating and verifying requests for interactive zones ('RELEASE' and 'COORDINATED_REPLANNING'), the state message is used by the vehicle to request an authorization and the MQTT topic `response` is used by master control to grant or revoke the authorization.

Author: ```suggestion
For communicating requests for the interactive zones 'RELEASE' and 'COORDINATED_REPLANNING', the field `zoneRequests` in the state message is used. The separate topic `response` is used by master control to respond to these requests.
```

> The authorization request is related to the vehicle state in which the request is embedded.

Author: ```suggestion
```

Author: This is misleading and should be clear

> For a 'RELEASE' zone, the vehicle requests access to the zone before entering the zone.

Author: ```suggestion
Before entering an interactive zone, the vehicle shall state a request.
```

> A request for entering the RELEASE zone is necessary, even if the order contains released nodes within the zone.

Author: ```suggestion
A request before entry of an interactive zone is necessary, even if the order contains released nodes within the zone.
```

> The vehicle decides at what distance to the zone it requests access.

Author: ```suggestion
The vehicle decides at which point before entering the zone to make its requests.
```

> If the access isn't granted in time, the vehicle shall not enter the zone.

Author: ```suggestion
If the response is not received in time, the vehicle shall not enter the zone.
```

> During traversing a 'COORDINATED_REPLANNING' zone, the vehicle can request a new path through the zone if the vehicle cannot proceed on its previously planned path.

Author: ```suggestion
```

> To request access or permission for a path change, the vehicle appends one or several zone requests related to interactive zones to its state message in the `zoneRequests` field.

Author: ```suggestion
```

Author: Stated above.

  | | directedLimitation | string | Enum {'SOFT','RESTRICTED','STRICT'}<\br>SOFT: Mobile robots may deviate from the defined direction of travel, but should avoid it, RESTRICTED: The mobile robot may deviate from the defined direction of travel, e.g., for obstacle avoidance, but shall never traverse opposite to the defined direction of travel, STRICT: The mobile robot shall maintain the defined direction of travel as precise as possible. |
| BIDIRECTED | | | While in this zone, vehicles shall only move in the defined direction of travel and its direct opposite (+ Pi), vehicles should not cross this zone in any other direction. | 
| | direction | float64 | Preferred direction of travel within the zone in radians. The direction of travel is the speed vector in the project-specific coordinate system. | 
| | bidirectedLimitation | string | Enum {'SOFT', 'RESTRICTED'}<\br>SOFT: Mobile robots may deviate from the defined directions of travel, but should avoid it, RESTRICTED: The mobile robot should not traverse in any other direction than the directions of travel, except for obstacle avoidance.|

## 6.x.2 Sharing of planned path for freely navigating mobile robots

> Vehicles shall communicate their planned trajectory to the master control system. This is done via the state message. For a higher frequency of sharing, the visualisation topic can be used. For this purpose the parameters `plannedPath` and `intermediatePath` are introduced, to be used only for trajectories planned by the mobile robot. The trajectory fields in the edgeState shall only be used to 'acknowledge' trajectories that have already been defined a priori within a layout or order and not during operations.

Author: ```suggestion
Vehicles shall communicate their planned trajectory to the master control system. This is done via the state message. For a higher frequency of sharing, the visualization topic can be used.
```

Author: Moved the sentence to end of chapter

This recommendation distinguishes between the `plannedPath`, which represents a longer path within the robot's currently active order that it is confident to share, and the `intermediatePath`, which represents the estimated time of arrival at closer waypoints that the vehicle is able to perceive with its sensors. Both paths start from the current position of the mobile robot (independent of any node that is part of the order the robot is currently executing) and go as far as the robot has what it considers to be reliable data, as it may also be situation dependent.

> The `plannedPath` is defined as a NURBS identical to the one defined in the `trajectory` field of the `edgeState`. In addition, the planned path may contain an array of nodes (referenced by their `nodeId`) that will be traversed as part of the current path.

Author: ```suggestion
The `plannedPath` is defined as NURBS as defined in the `trajectory` field of the `edgeState`. The planned path can contain an array of nodes, referenced by their `nodeId`, that will be traversed as part of the current path.
```

> The `intermediatePath` is defined as a polyline with an estimated time of arrival defined by a `timestamp` and optionally the orientation of the mobile robot added to each shared waypoint. It is defined as an array of waypoints where the line segments of a `polyline` are linear. The exact length of the shared path is limited by the robot's perception of its intermediate environment by its sensors.

Author: ```suggestion
The `intermediatePath` is defined as a polyline. The polyline consists of linear line segments between waypoints. Each `waypoint` consists of its `x` and `y` position, an optional orientation of the vehicle and the `ETA` indicating the estimated time of arrival.
```

> 

Author: ```suggestion
The parameters `plannedPath` and `intermediatePath` shall be used only for trajectories planned by the mobile robot. The trajectory fields in the edgeState shall only be used to 'acknowledge' trajectories that have already been defined a priori within a layout or the order.
 
```

The master control can request the deletion of a specific map from a vehicle. This is done with the instant action `deleteMap`. When a vehicle runs out of memory, it should report this to the master control, which can then initiate the deletion of maps. The vehicle itself is not allowed to delete maps.
After successfully deleting a map, it is important to remove that map's entry from the vehicle's array of maps in the vehicle state.

## 6.2 Sharing of planned path for freely navigating mobile robots (Standalone chapter)

> Vehicles shall communicate their planned trajectory to the master control system. This is done via the state message. For a higher frequency of sharing, the visualisation topic can be used. For this purpose the parameters `plannedPath` and `intermediatePath` are introduced, to be used only for trajectories planned by the mobile robot. The trajectory fields in the edgeState shall only be used to 'acknowledge' trajectories that have already been defined a priori within a layout or order and not during operations.

Author: ```suggestion
Vehicles shall communicate their planned trajectory to the master control system. This is done via the state message. For a higher frequency of sharing, the visualisation topic can be used. For this purpose the parameters `plannedPath` and `intermediatePath` are introduced, to be used only for trajectories planned by the mobile robot. The trajectory fields in the edgeState shall only be used to acknowledge trajectories that have already been defined a priori within a layout or order and not during operations.
```

## 6.x Zones

Zones are used to define rules for specific areas of the vehicle workspace. In this way, zones allow vehicles to navigate freely between nodes while giving the Master Control the ability to manage traffic. Zones can be used to locally deny vehicles access to areas or to link access to conditions (zone types: 'BLOCKED' and 'RELEASE'). It is also possible to enforce specific behavior while within the zone (zone types: 'LINE_GUIDED', 'SPEED_LIMIT', 'COORDINATED_REPLANNING', and 'ACTION') or influence the driving behavior by incentivizing or penalizing certain areas (zone types: 'PRIORITY' and 'PENALTY') or giving a predefined driving direction (zone type: 'DIRECTED', 'BIDIRECTED'). The zone types are defined in the following sections.

Potential conflicts in orders due to overlapping of zones or combination of zone and edge properties and how to resolve them are addressed in section 10.4.

> Some vehicles cannot process zones at all, while other vehicles might only be able to work with certain more basic zone types like 'BLOCKED'. All vehicles must therefore report to master control which zones they are able to understand by adding the according zone names to the `supportedZones` array under `typeSpecifications` in their factsheet.

Reviewer 2: ```suggestion
Some vehicles cannot process zones at all, while other vehicles might only be able to work with a certain subset of zone types e.g.  'BLOCKED'. All vehicles must therefore report to master control which zones they are able to understand by adding the according zone names to the `supportedZones` array under `typeSpecifications` in their factsheet.
```

Author: ```suggestion
Some vehicles cannot process zones at all, while other vehicles might only be able to work with a certain subset of zone types, such as 'BLOCKED'. All vehicles must therefore report to master control which zones they are able to understand by adding the according zone names to the `supportedZones` array under `typeSpecifications` in their factsheet.
```

Also (virtually) line-guided vehicles can choose to support zone-based navigation if they can implement the logic of the corresponding zone types defined in the following. 

> If a certain behavior can be achieved by one of the defined zone types or a combination thereof, these shall be used and no proprietary versions shall be introduced.

Author: ```suggestion
```

Author: Not necessary, no custom zone sets allowed.

> If a certain behavior can be archived by one of the defined zone types or a combination thereof, these shall be used and no proprietary versions shall be introduced.

Author: ```suggestion
If a certain behavior can be achieved by one of the defined zone types or a combination thereof, these shall be used and no proprietary versions shall be introduced.
```

> In its state message, each vehicles reports in the `activeZoneSetId` field for each `map` object in the `maps` array the `zoneSetId` that is currently in active use. There shall be no references to map or zone IDs that are not on the vehicle.

Reviewer 2: ```suggestion
In its state message, each vehicles reports in the `activeZoneSetId` field for each `map` object in the `maps` array the `zoneSetId` that is currently in active use. There shall be no references to a mapId or zoneID that is not on the vehicle.
```

Author: ```suggestion
```

A zone set shall only be changed and distributed by master control to keep consistency in the system.

### 6.x.1 Zone types

Two types of zones are distinguished: contour-based zones and kinematic center-based zones. This distinction is based on the different conditions for when the vehicle is considered to be entering and exiting zones.

#### Contour-based zones

> For contour-based zones, the vehicles contour (including its load) decides the entry and exit of the zones. These zones shall not be entered by any part of the contour, if the master control does not allow it, e.g., a 'BLOCKED' zone may never be entered, a 'RELEASE' zone may only be entered, if it is released by the master control, etc. or they lead to a defined behavior when entered, e.g., in the ACTION zone when entering all entryActions shall be executed and so on.

Reviewer 2: ```suggestion
For contour-based zones, the vehicles contour (including its load) decides the entry and exit of the zones. Any part of the contour entering the zone is considered a zone entry whereas an exit is only then completed when no part of the contour is within the zone anymore. So, if the master control does not allow it, e.g., a 'BLOCKED' zone is not be entered by any part of the contour, a 'RELEASE' zone may only be entered, if it is released by the master control, etc. or they lead to a defined behavior when entered, e.g., in the ACTION zone when entering all entryActions shall be executed and so on.
```

Author: ```suggestion
For contour-based zones, the contour of the vehicle (including its load) determines zone entry and exit. Any part of the contour entering the zone is considered a zone entry. A zone exit shall be reported only when no part of the contour is inside the zone. When a zone is reported as exited is decided by the mobile robot. 
```

Author: Make description more precise, removed comments on zones, that are later introduced, reducing redundancy.

![Figure x1 Depiction of a vehicle entering a zone based on its contour (left) and a loaded vehicle with corresponding extended bounding box exiting a zone (right)](./assets/contour_entry.png)
>Figure x1 Depiction of a vehicle entering a zone based on its contour (left) and a loaded vehicle with corresponding extended bounding box exiting a zone (right)

The following contour-based zones are defined:

| **Zone Type**| **Zone Parameters** | **Data type** | **Description** | 
| --- | --- | --- | --- |

> | BLOCKED | none | - | Vehicles shall not enter this zone. If a vehcile has entered the zone, it shall stop and throw a FATAl error.| 

Reviewer 2: ```suggestion
| BLOCKED | none | | Vehicles shall not enter this zone. If a vehicle has entered the zone or finds itself within one, it shall stop and throw a FATAL error.| 
```

> | LINE_GUIDED | none | - | No free navigation is allowed in this zone, mobile robots shall follow the defined trajectories on edges. Vehicles may only enter this zone if the route is explicitly specified by the master control in the form of a node-edge graph. When entering the zone, the mobile robot shall be on the trajectory of the edge between the last node outside the zone and the first node inside the zone on the edge when entering. The edges that enter and are within the line-guided zone require a trajectory sent from the master control or a predefined trajectory on the vehicle. A corridor can be sent to allow the vehicle to deviate from their trajectory. | 

Reviewer 2: ```suggestion
| LINE_GUIDED | none | | No free navigation is allowed in this zone, mobile robots shall follow the defined trajectories on edges. Vehicles may only enter this zone if the route is explicitly specified by the master control in the form of a node-edge graph. Any movement of the mobile robot that requires it to enter this zone shall follow a predefined trajectory. When entering the zone, the mobile robot shall be on the trajectory of the edge that crosses the zone. The edges that enter and are inside the line-guided zone require a trajectory sent from the master control or a predefined trajectory on the vehicle. A corridor can be sent to allow the vehicle to deviate from the trajectory. | 
```

Reviewer 2: probably not necessary

Reviewer 2: Leerzeichen noch zu viel

| RELEASE | | - | Vehicles are only allowed entering this zone once they have been granted access through master control. | 

> | | releaseLossBehavior | enum | When the access to this zone is revoked or expired, the vehicle can either STOP, CONTINUE, or EVACUATE the zone. This Action is only executed, when the vehicle is already in the zone and the release expires or is revoked. If not defined, the vehicle is expected to STOP and report an error. STOP: Vehicle stops and sends a fatal error. EVACUATE: Execute the evacuation behavior of the vehicle to leave the zone, keeping the release in its state until the zone is left. CONTINUE: If the revoke/expiry happens, after the vehicle entered the zone, the vehicle continues its path keeping the reservation of the vehicle in its state. If the order ends inside the zone, the vehicle waits for a new order. | 

Reviewer 2: ```suggestion
| | releaseLossBehavior | string | Enum {'STOP', 'CONTINUE', 'EVACUATE'}</br>When the access to this zone is revoked or expired, the vehicle can either STOP, CONTINUE, or EVACUATE the zone. This action is only executed, when the vehicle is already in the zone and the release expires or is revoked. If not defined, the vehicle is expected to STOP and report an error. STOP: Vehicle stops and sends a FATAL error. EVACUATE: Execute the evacuation behavior of the vehicle to leave the zone, keeping the release in its state until the zone is left. CONTINUE: If the revoke/expiry happens, after the vehicle entered the zone, the vehicle continues its path keeping the release of the zone in its state. If the order ends inside the zone, the vehicle waits for a new order. | 
```

> | COORDINATED_REPLANNING | - | - | No autonomous replanning is allowed within this zone. Vehicles are only allowed adjusting their global or local path if granted by master control. | 

Reviewer 2: ```suggestion
| COORDINATED_REPLANNING | none | | No autonomous replanning is allowed within this zone. Vehicles are only allowed adjusting their path if granted permission by master control. | 
```

> | SPEED_LIMIT | | - | Vehicles must not drive faster than the defined maximum speed within this zone. | 

Author: ```suggestion
| SPEED_LIMIT | | | Vehicles must not drive faster than the defined maximum speed within this zone. | 
```

> | | maxSpeed | float64 | Maximum permitted speed for vehicles within the zone in m/s. The speed limit shall be reached when entering the zone.|

Reviewer 2: ```suggestion
| | maxSpeed | float64 | Maximum permitted speed for vehicles within the zone in m/s. The speed limit shall already be reached upon entering the zone.|
```

> | ACTION | | - | Vehicles shall perform predefined actions when entering, traversing, or exiting the zone. The factsheet defines which actions can be executed when. |

Author: ```suggestion
| ACTION | | | Vehicles shall perform predefined actions when entering, traversing, or exiting the zone. The factsheet defines which actions can be executed when. |
```

| | entryActions[action] | array | Actions to be triggered when entering the zone. Empty Array, if no Actions required. | 
| | duringActions[action] | array | Actions to be executed while crossing the zone. Empty Array, if no Actions required. | 
| | exitActions[action] | array | Actions to be triggered when leaving the zone. Empty Array, if no Actions required. | 

#### Kinematic center-based zones

In kinematic center-based zones, the vehicle's kinematic center decides the entry and exit of the zones. When the vehicle's kinematic center is within a zone, the vehicle shall follow the defined behavior. 

> PRIORITY and PENALTY zones are zones which only influence the path planning of mobile robots.

Reviewer 2: ```suggestion
'PRIORITY' and 'PENALTY' zones are zones which only influence the path planning of mobile robots.
```

> 'DIRECTED' zones define a preferred direction of travel within the zone. 'BIDIRECTED' zones define a travel direction and its opposite direction to be used.  Other directions shall be avoided. Additional parameters indicate the limits within which the vehicle may deviate from its direction of travel. The direction of travel is the speed vector in the project specific coordinate system.

Reviewer 2: ```suggestion
'DIRECTED' zones define a preferred direction of travel within the zone. 'BIDIRECTED' zones define a travel direction and its opposite direction to be used. Other directions shall be avoided. Additional parameters indicate the limits within which the vehicle may deviate from its direction of travel. The direction of travel is the speed vector in the project-specific coordinate system.
```

Reviewer 2: Leerzeichen

![Figure x2 Depiction of a vehicle entering a zone based on its kinematic center (left) and a loaded vehicle exiting a zone based on its kinematic center (right)](./assets/kinematic_center_entry.png)
>Figure x2 Depiction of a vehicle entering a zone based on its kinematic center (left) and a loaded vehicle exiting a zone based on its kinematic center (right)

| **Zone Type**| **Zone Parameters** | **Data type** | **Description** | 
| --- | --- | --- | --- |

> | PRIORITY | | | Navigating through this zone is associated with a reduced cost for the vehicle's route compared to an otherwise equivalent area with no such zone on the map. This zone is intended to incentivise use by vehicles.| 

Reviewer 2: ```suggestion
| PRIORITY | | | Navigating through this zone is associated with an incentivisation for the vehicle's route compared to an otherwise equivalent area with no such zone on the map.| 
```

Reviewer 2: isnt directly reduced cost but more a reward? Maybe reward instead of incentivisation.

| | priorityFactor | float64 | [0.0...1.0]<br> Relative factor, determining the zone's preference over an area with no zone. 0.0 means no preference, as if there was no zone, 1.0 is maximum preference. |
| PENALTY | | | Navigating through this zone is associated with an increased cost for the vehicle's route compared to an otherwise equivalent area with no such zone on the map. This zone is intended to decentivise use by vehicles.| 
| | penaltyFactor | float64 | [0.0...1.0]<br> Relative factor, determining the zone's penalty over an area with no zone. 0.0 means no penalty, as if there was no zone, 1.0 is the maximum penalty, causing the mobile robot to take this path only if no other path is possible.  |

> | DIRECTED | | - | Vehicles shall traverse this zone in specific directions of travel. | 

Reviewer 2: ```suggestion
| DIRECTED | | | Vehicles shall traverse this zone in a specific direction of travel. | 
```

| | direction | float64 | Preferred direction of travel within the zone in radians. The direction of travel is the speed vector in the project-specific coordinate system. | 

> | | limitation | enum | ['SOFT','RESTRICTED','STRICT']<\br>SOFT: Mobile robots may deviate from the defined direction of travel, but should avoid it, RESTRICTED: The mobile robot may deviate from the defined direction of travel, e.g., for obstacle avoidance, but shall never traverse opposite to the defined direction of travel, STRICT: The mobile robot shall not deviate from the defined direction of travel (within its precision). |

Reviewer 2: ```suggestion
| | limitation | enum | ['SOFT','RESTRICTED','STRICT']<\br>SOFT: Mobile robots may deviate from the defined direction of travel, but should avoid it, RESTRICTED: The mobile robot may deviate from the defined direction of travel, e.g., for obstacle avoidance, but shall never traverse opposite to the defined direction of travel, STRICT: The mobile robot shall maintain the defined direction of travel as precise as possible. |
```

Author: Other rework

> | | limitation | enum | ['SOFT','RESTRICTED','STRICT']<\br>SOFT: Mobile robots may deviate from the defined direction of travel, but should avoid it, RESTRICTED: The mobile robot may deviate from the defined direction of travel, e.g., for obstacle avoidance, but shall never traverse opposite to the defined direction of travel, STRICT: The mobile robot shall not deviate from the defined direction of travel (within its precision). |

Author: ```suggestion
  | | directedLimitation | string | Enum {'SOFT','RESTRICTED','STRICT'}<\br>SOFT: Mobile robots may deviate from the defined direction of travel, but should avoid it, RESTRICTED: The mobile robot may deviate from the defined direction of travel, e.g., for obstacle avoidance, but shall never traverse opposite to the defined direction of travel, STRICT: The mobile robot shall maintain the defined direction of travel as precise as possible. |
```

> | BIDIRECTED | | | While in this zone, vehicles shall only move in the direction of travel and the opposite (+ Pi), vehicles should not cross this zone in any other direction. | 

Reviewer 2: ```suggestion
| BIDIRECTED | | | While in this zone, vehicles shall only move in the defined direction of travel and its direct opposite (+ Pi), vehicles should not cross this zone in any other direction. | 
```

> | | direction | float64 | Direction and its opposite direction (+ Pi) define the allowed directions of travel within the zone in radians. The direction of travel is the speed vector in the project-specific coordinate system. | 

Reviewer 2: "Direction (and the implicitly defined direct opposite direction (+Pi))...."?

> | | direction | float64 | Direction and its opposite direction (+ Pi) define the allowed directions of travel within the zone in radians. The direction of travel is the speed vector in the project-specific coordinate system. | 

Author: ```suggestion
| | direction | float64 | Preferred direction of travel within the zone in radians. The direction of travel is the speed vector in the project-specific coordinate system. | 
```

Author: Idea: Same definition for directed and bidirected, since same parameter (both cases just a direction) Implicitly understand to use different when in BIDIRECTED zone.

> | | limitation | enum | ['SOFT', 'RESTRICTED']<\br>SOFT: Mobile robots may deviate from the defined directions of travel, but should avoid it, RESTRICTED: The mobile robot should not traverse in any other direction than the directions of travel, exept for obstacle avoidance.|

Reviewer 2: ```suggestion
| | bidirectedLimitation | string | Enum {'SOFT', 'RESTRICTED'}<\br>SOFT: Mobile robots may deviate from the defined directions of travel, but should avoid it, RESTRICTED: The mobile robot should not traverse in any other direction than the directions of travel, except for obstacle avoidance.|
```

## 6.x.2 Sharing of planned path for freely navigating mobile robots

> Vehicles shall communicate their planned trajectory to the master control system. This is done via the state message. For a higher frequency of sharing, the visualisation topic can be used. For this purpose the parameters `plannedPath` and `intermediatePath` are introduced, to be used only for trajectories planned by the mobile robot. The trajectory fields in the edgeState shall only be used to 'acknowledge' trajectories that have already been defined a priori within a layout or order and not during operations.

Reviewer 2: ```suggestion
Vehicles shall communicate their planned trajectory to the master control system. This is done via the state message. For a higher frequency of sharing, the visualization topic can be used. The parameters `plannedPath` and `intermediatePath` provided for this, are only to be used for trajectories planned by the mobile robot. The trajectory fields in the edgeState shall only be used to 'acknowledge' trajectories that have already been defined a priori within a layout or order and not during operations.
```

Reviewer 2: "introdeced" might sound outdated after a while

> This recommendation distinguishes between the `plannedPath`, which represents a longer path within the robot's currently active order that it is confident to share, and the `intermediatePath`, which represents the estimated time of arrival at closer waypoints that the vehicle is able to perceive with its sensors. Both paths start from the current position of the mobile robot (independent of any node that is part of the order the robot is currently executing) and go as far as the robot has what it considers to be reliable data, as it may also be situation dependent.

Reviewer 2: ```suggestion
Vehicles share their `intermediatePath`, which represents the estimated time of arrival at closer waypoints that the vehicle is able to perceive with its sensors, and their `plannedPath`, which represents a longer path within the robot's currently active order. Both paths shall start from the mobile robot's current position, independent of any nodes that are part of the order. The robot can decide on the length of the shared paths, as it may be situation dependent.
```

The `plannedPath` is defined as a NURBS identical to the one defined in the `trajectory` field of the `edgeState`. In addition, the planned path may contain an array of nodes (referenced by their `nodeId`) that will be traversed as part of the current path.

> The `intermediatePath` is defined as a polyline with an estimated time of arrival defined by a `timestamp` and optionally the orientation of the mobile robot added to each shared waypoint. It is defined as an array of waypoints where the line segments of a `polyline` are linear. The exact length of the shared path is limited by the robot's perception of its intermediate environment by its sensors.

Reviewer 2: ```suggestion
The `intermediatePath` is defined as a polyline with an estimated time of arrival defined by a `timestamp` and optionally the orientation of the mobile robot added to each shared waypoint. It is defined as an array of waypoints where the line segments of a `polyline` are linear. The exact length of the shared path is limited by what the robot's sensors perceive of its intermediate environment.
```

### 6.x. Implementation of the zone set object for zone set transfer

In contrast to the maps, the structure of the zones is predefined. This structure shall be maintained when providing the download as well as sending the zones via the separate topic `zoneSet`. It is not part of the state message of the vehicles.

> A zoneSet shall only be changed and distributed by master control to keep consistency in the system.

Reviewer 2: ```suggestion
Zone sets shall only be changed and distributed by master control to keep consistency in the system.
```

> A `zoneSet` is an array of `zone` objects with a globally unique identifier, `zoneSetId`. It must be associated with a single map referenced through the `mapId`. The `mapVersion` shall not be referenced, as the same zone set might be intended to be used for several versions of one map. In general, several zone sets can be defined in addition to a single map and it is upon master control to ensure that the right zone set is enabled for each map on the vehicle. As with maps, the `zoneSetStatus` indicates which zone sets is currently used by the device. Only a single zone set can be activated at a time for each `mapId` on the vehicle. Zones shall not extend beyond the spatial boundaries of a map.

Reviewer 2: ```suggestion
A `zoneSet` is an array of `zone` objects with a globally unique identifier, `zoneSetId`. It must be associated with a single map referenced through the `mapId`. The `mapVersion` shall not be referenced, as the same zone set might be intended to be used for several versions of one map. In general, several zone sets can be defined in addition to a single map and it is upon master control to ensure that the right zone set is enabled for each map on the vehicle. As with maps, the `zoneSetStatus` indicates which zone set is currently used by the vehicle. Only a single zone set can be active at once for each `mapId` on the vehicle. Zones shall not extend beyond the spatial boundaries of a map.
```

The content of a zone set with a unique zoneSetId shall not change. If changes are required within a zone set, it shall be identified by a new zoneSetId.

| **Object structure** | **Data type** | **Description** |
| --- | --- | --- |
| zoneSet{ | JSON object | Zone set detailing a dedicated map. | 
| mapId | string | Globally unique identifier of the map the zone set particularizes. | 
| zoneSetId | string | Globally unique identifier of the zone set. |
| *zoneSetDescription* | string | Human readable description of the zone set. | 
| zones[zone] <br> } | array | Array of zone objects. | 

A single zone object has the following structure: 

| **Object structure** | **Data type** | **Description** |
| --------------------- | ------------- | ------------------- |
| zone{ | JSON object | |
| zoneId | string | Locally (within the zone set) unique identifier. |
| zoneType | enum | Enum {'BLOCKED', 'LINE_GUIDED', 'RELEASE', 'COORDINATED_REPLANNING', 'SPEED_LIMIT', 'ACTION', 'PRIORITY', 'PENALTY', 'DIRECTED', 'BIDIRECTED'}, Zone type according to Section 6.x.1 Zone types. |
| *zoneDescription* | string | User-defined human-readable name or descriptor. | 

> | **vertices[vertex]** <br> } | array | Array of vertices (in x-y-coordinates) defining the geometrical shape of the zone clockwise. |

Reviewer 2: ```suggestion
| **vertices[vertex]**| array | Array of vertices that define the geometric shape of the zone in a clockwise direction. |
```

| *maxSpeed* | float64 | Required in SPEED_LIMIT zone as defined in chapter 6.x.1.| 
| ***entryActions[Action]*** | array | Required in ACTION zone as defined in chapter 6.x.1.| 
| ***duringActions[Action]*** | array | Required in ACTION zone as defined in chapter 6.x.1.| 
| ***exitActions[Action]*** | array | Required in ACTION zone as defined in chapter 6.x.1.| 
| *releaseLossBehavior* | enum | Optional for a RELEASE zone as defined in chapter 6.x.1.|

> The shape of each zone object is defined through a polygon, which is communicated through its vertices. A minimum of three vertices must be defined to make-up a full polygon. If the first entry of the array of vertices is not identical to the last, implicitly the polygon is closed through a connection line to the first vertex. Only simple (meaning without any intersections) polygons are supported. The array of vertices defining a zone is provided as a list of x-y tuples in the globally defined project-specific coordinate system in a clockwise direction: 

Reviewer 2: ```suggestion
The shape of each zone object is defined through a polygon, which is communicated through its vertices. A zone with less than three vertices is invalid and shall be rejected. If the first entry of the vertex array is not identical to the last, the polygon is implicitly closed by a connecting line to the first vertex. Only simple polygons (i.e. without intersections) shall be used. The array of vertices defining a zone is provided as a list of x-y tuples in the globally defined project-specific coordinate system in a clockwise direction: 
```

Reviewer 2: or make it that if its less than 3 vertices its invalid?

| **Object structure** | **Data type** | **Description** |
| --------------------- | ------------- | ------------------- |
| vertex{| JSON-object| |

> | x | float64 | X-coordinate described in the project specific coordinate system |

Reviewer 2: ```suggestion
| x | float64 | X-coordinate described in the project-specific coordinate system |
```

> | y <br>} | float64 | Y-coordinate described in the project specific coordinate system |

Reviewer 2: ```suggestion
| y <br>} | float64 | Y-coordinate described in the project-specific coordinate system |
```

## 6.x.3 Interactive zones - Communication 

For communicating and verifying requests for interactive zones ('RELEASE' and 'COORDINATED_REPLANNING'), the state message is used by the vehicle to request an authorization and the MQTT topic `response` is used by master control to grant or revoke the authorization.
The authorization request is related to the vehicle state in which the request is embedded.

For a 'RELEASE' zone, the vehicle requests access to the zone before entering the zone.
A request for entering the RELEASE zone is necessary, even if the order contains released nodes within the zone.

> The vehicle decides in which distance the access is requested.

Reviewer 2: ```suggestion
The vehicle decides at what distance to the zone it requests access.
```

If the access isn't granted in time, the vehicle shall not enter the zone.

During traversing a 'COORDINATED_REPLANNING' zone, the vehicle can request a new path through the zone if the vehicle cannot proceed on its previously planned path.

> To request access or permission for a path change, the vehicle appends one or several zone requests related to interactive zones to its state message in `zoneRequests` array.

Reviewer 2: ```suggestion
To request access or permission for a path change, the vehicle appends one or several zone requests related to interactive zones to its state message in the `zoneRequests` field.
```

Only requests to zones of enabled zone sets are valid. Zone requests can be made to zones that belong to maps that the vehicle is currently not on.
If the vehicle requests access to a zone or permission for replanning in a zone within a zone set that is not enabled, master control shall always reject the request.

> The `requestID` allows master control to differentiate between different requests leading to new requests if the vehicle got denied once and allows the vehicle to request multiple alternatives at the same time.

Reviewer 2: ```suggestion
The `requestId` allows master control to distinguish between different requests and allows the vehicle to request multiple alternatives at the same time.
```

Reviewer 2: consistent with other naming of Id parameters

Reviewer 2: also the sentence might be confusing

Each request shall use an unique identifier per robot. Ids can be reused after a vehicle reboot.

For requests related to a 'RELEASE' zone, a `zoneRequest` object of `requestType` 'ACCESS' shall be added to the state message.
For permission to enter a 'COORINATED_REPLANNING' zone with a planned path or for replanning its path within the this zone, the request type shall be set to 'REPLANNING'.
Request-specific parameters are added as optional parameters to the request.

> For 'REPLANNING' request, the `trajectory` parameter is added to describe the requested path as type of trajectory from the state message.

Reviewer 2: ```suggestion
For a 'REPLANNING' request, the planned path shall be added to the `trajectory` field.
```

Reviewer 2: "as type of trajectory from the state message" sounds confusing

|timestamp | | string | Timestamp (ISO 8601, UTC); YYYY-MM-DDTHH:mm:ss.fffZ (e.g., "2017-04-15T11:40:03.123Z").
|version | | string | Version of the protocol [Major].[Minor].[Patch] (e.g., 1.3.2).
|manufacturer | | string | Manufacturer of the AGV.
|serialNumber | | string | Serial number of the AGV.
|responses[response] | array | Array of response objects. |

Object structure/Identifier | Data type | Description
| --- | --- | --- |
| response <br> { | JSON object | Object which contains the master control answer to a specific request. |
| requestId | string | Unique per vehicle identifier within all active requests. |
| grantType | enum | Enum {'GRANTED','QUEUED','REVOKED','REJECTED'}<\br>'GRANTED': master control grants request. 'REVOKED': master control revokes previously granted request. 'REJECTED': master control rejects a request. 'QUEUED': Acknowledge the vehicles request to the master control, but no permission is given yet. Request was added to some sort of a queue. |
| *leaseExpiry* | string | Timestamp (ISO 8601, UTC); YYYY-MM-DDTHH:mm:ss.fffZ (e.g.“2017-04-15T11:40:03.123Z”)

### 6.x.4 Potentially conflicts in orders for vehicles working with zones

> In the following matrix possible interactions between zones are described. The matrix is symmetric, as the interaction between two zones is the same, regardless of the order in which they are considered. For each combination, there is either a zone behavior that is overrulling the other (e.g., a 'BLOCKED' zone overrules a 'LINE_GUIDED' zone) or there is no conflict (e.g., a 'LINE_GUIDED' zone and a 'COORDINATED_REPLANNING' zone). 'DIRECTED' and 'BIDIRECTED' zones shall not overlap, since this might lead to an undefined behavior. The Column No Zone defines the behavior for contour-based zones, where vehicles can be inside a defined zone type and an area without a zone at the same time. For kinematic center-based zones the vehicle can only be completely within or outside  the zone, so there is no interaction existing.

Reviewer 2: ```suggestion
In the following matrix possible interactions between zones are described. The matrix is symmetric, as the interaction between two zones is the same, regardless of the order in which they are considered. For each combination, there is either a zone behavior that is overrulling the other (e.g., a 'BLOCKED' zone overrules a 'LINE_GUIDED' zone) or there is no conflict (e.g., a 'LINE_GUIDED' zone and a 'COORDINATED_REPLANNING' zone). 'DIRECTED' and 'BIDIRECTED' zones shall not overlap, since this might lead to an undefined behavior. The column No Zone defines the behavior for contour-based zones, where vehicles can be inside a defined zone type and an area without a zone at the same time. For kinematic center-based zones the vehicle can only be completely within or outside the zone, so there is no possible interaction.
```

---|---|---|---|---|---|---|---|---|---|---|---|---
**BLOCKED**|BLOCKED|BLOCKED|BLOCKED|BLOCKED|BLOCKED|BLOCKED|BLOCKED|BLOCKED|BLOCKED|BLOCKED|BLOCKED|BLOCKED|
**RELEASE**||No Conflict|No Conflict|No Conflict|No Conflict|No Conflict|No Conflict|No Conflict|No Conflict|No Conflict|No Conflict|No Conflict
**LINE_GUIDED**|||No conflict|LINE_GUIDED|No Conflict| (1) |LINE_GUIDED|LINE_GUIDED|LINE_GUIDED|No conflict|LINE_GUIDED|No conflict
**COORDINATED_REPLANNING**||||(2)|No conflict|(1)|No conflict|No conflict|No conflict|No conflict|COORDINATED_REPLANNING|(3)
**SPEED_LIMIT** |||||(4)|No conflict|No conflict|No conflict|No conflict|No conflict|SPEED_LIMIT|(4)
**ACTION** ||||||(5)|No conflict|No conflict|No conflict|No conflict|ACTION|(5)
**PRIORITY** |||||||(6)|(6)|No conflict|No conflict|(7)|No conflict
**PENALTY** ||||||||(6)|No conflict|No conflict|(7)|No conflict
**DIRECTED** |||||||||(8)|(8)|(7)|(9)
**BIDIRECTED** ||||||||||(8)|(7)|(9)

1) If actions would conflict with other zones behavior, report a CRITICAL error (order error) and stop the vehicle.

> 2) Planned trajectory must be granted for all COORDINATED_REPLANNING zones.

Reviewer 2: ```suggestion
2) Planned trajectory must be granted for all 'COORDINATED_REPLANNING' zones.
```

3) If a trajectory is predefined for the edge, it shall be sent in the zone request.

> 4) The lower of the two competing speeds applies.

Reviewer 2: ```suggestion
4) The lowest of the competing speeds applies.
```

5) Execute all actions

> 6) The more restrictive one is always selected here; for two PRIORITY zones, the lower priority factor is used; for one PRIORITY and one PENALTY zone, the penalty factor is used; for two PENALTY zones, the higher penalty factor is used.

Reviewer 2: ```suggestion
6) The most restrictive one is always selected here; for PRIORITY zones, the lowest priority factor is used; for overlapping PRIORITY and PENALTY zones, the highest penalty factor is used; for overlapping PENALTY zones, the highest penalty factor is used.
```

Reviewer 2: should apply to x-amount of overlapping zones right?

***ZoneSets[ZoneSet]*** | | Array of ZoneSet | Array of ZoneSet objects that are currently stored on the vehicle.
orderId| | string | Unique order identification of the current order or the previously finished order. <br>The orderId is kept until a new order is received. <br>Empty string (""), if no previous orderId is available.
orderUpdateId | | uint32 | Order update identification to identify, that an order update has been accepted by the AGV. <br>"0" if no previous orderUpdateId is available.
lastNodeId | | string | Node ID of last reached node or, if the AGV is currently on a node, current node (e.g., "node7"). Empty string (""), if no `lastNodeId` is available.
lastNodeSequenceId | | uint32 | Sequence ID of the last reached node or, if the AGV is currently on a node, Sequence ID of current node. <br>"0" if no `lastNodeSequenceId` is available.
**nodeStates [nodeState]** | |array | Array of nodeState objects that need to be traversed for fulfilling the order<br>(empty array if idle)
**edgeStates [edgeState]** | |array | Array of edgeState objects that need to be traversed for fulfilling the order<br>(empty array if idle)

> ***plannedPath*** | | JSON object |  

Author: ```suggestion
***plannedPath*** | | JSON object |  Represents a path within the robot's currently active order as NURBS.
```

> ***intermediatePath*** | | JSON object |  

Author: ```suggestion
***intermediatePath*** | | JSON object |  Represents the estimated time of arrival at closer waypoints that the vehicle is able to perceive with its sensors.
```

mapVersion | | string | Version of the map.
mapStatus <br>}| | string | Enum {'ENABLED', 'DISABLED'}<br>'ENABLED': Indicates this map is currently active / used on the AGV. At most one map with the same mapId can have its status set to 'ENABLED'.<br>'DISABLED': Indicates this map version is currently not enabled on the AGV and thus could be enabled or deleted by request.

Object structure | Unit | Data type | Description
---|---|---|---
**ZoneSet**{ | | JSON object|
zoneSetId | | string | Unique identifier of the zone set that is currently enabled for the map.<br> This field shall be left empty only if the vehicle has no zones defined for the corresponding map.

> mapId | | string | 

Author: ```suggestion
mapId | | string | Identifier of the corresponding map.
```

} | | |

Object structure | Unit | Data type | Description
 ---|---|---|---
 **trajectory** { | | JSON object |
 *degree* | | uint32 | Degree of the NURBS curve defining the trajectory.<br><br>Range: [1 ... uint32.max]<br>Default: 1
 ***knotVector [float64]*** | | array | Array of knot values of the NURBS.<br>The size of `knotVector` is exactly `degree` + 1 larger than the size of `controlPoints`.<br>The multiplicities of the first and last knot, both, must be `degree` + 1 (clamped NURBS).<br>The multiplicity of knots other than the first or last knot must not be greater than `degree` (continuity).<br><br>Range of knots: [0.0 ... 1.0]<br>Default: Equidistant knots from 0.0 to 1.0 with a multiplicity of `degree` + 1 for the first and last knot, and multiplicity 1 for all other knots (uniform knots).
 **controlPoints [controlPoint]** | | array | Array of controlPoint objects defining the control points of the NURBS, explicitly including the start and end point (clamped NURBS).<br>The number of control points needs to be at least `degree` + 1.
 } | | |

Object structure | Unit | Data type | Description
---|---|---|---
**controlPoint** { | | JSON object |

> x | m | float64 | X-coordinate described in the world coordinate system.

Reviewer 3: ```suggestion
x | m | float64 | X-coordinate described in the project specific coordinate system.
```

> y | m | float64 | Y-coordinate described in the world coordinate system.

Reviewer 3: ```suggestion
y | m | float64 | Y-coordinate described in the project specific coordinate system.
```

*weight* | | float64 | The weight of the control point on the curve.<br><br>Range: ]0.0 ... float64.max]<br>Default: 1.0
} | | |

Object structure | Unit | Data type | Description
---|---|---|---
**intermediatePath** { | | JSON object |  
**polyline[waypoint]** | | array | Array of end points of segments of a polyline. 
} | | |

Object structure | Unit | Data type | Description
---|---|---|---
**waypoint** { | | JSON object | Endpoint of a segment within a defined polyline.

> x | m | float64 | X-coordinate described in the world coordinate system.

Reviewer 3: ```suggestion
x | m | float64 | X-coordinate described in the project specific coordinate system.
```

> y | m | float64 | Y-coordinate described in the world coordinate system.

Reviewer 3: ```suggestion
y | m | float64 | Y-coordinate described in the project specific coordinate system.
```

> *theta* | rad | float64 | Absolute orientation of the vehicle in the world coordinate system. <br> Range: [-Pi ... Pi] </br>

Reviewer 3: ```suggestion
*theta* | rad | float64 | Absolute orientation of the vehicle in the project specific coordinate system. <br> Range: [-Pi ... Pi] </br>
```

***loadDimensions*** | | JSON object | Dimensions of the load's bounding box in meters.
*weight*<br>} | kg | float64 | Range: [0.0 ... float64.max]<br><br>Absolute weight of the load measured in kg.

| **Object structure** | **Data type** | **Description** |
| --- | --- | --- |
| zoneRequest <br> { | JSON object | Zone information sent by the vehicle to master control. |
| requestId | string | Unique per vehicle identifier within all active requests. |

> | requestType | enum | Enum specifying the type of zone the request relates to. Feasible values are ACCESS or REPLANNING. |

Reviewer 2: ```suggestion
| requestType | string | Enum {ACCESS, REAPLNNING} </br> Specifying the type of zone the request relates to. Feasible values are ACCESS or REPLANNING. |
```

| zoneId | string | Locally (within the zone set) unique identifier referencing the zone the request is related to. |
| zoneSetId | string | Due to the zoneId only being unique to a zoneSet, the zoneSetId is part of the request. |

> | requestStatus | enum | Enum {'REQUESTED', 'GRANTED', 'REVOKED', 'EXPIRED'}<br>When stating a request, this is set to REQUESTED. After response or update from master control set to GRANTED or REVOKED. If lease time expires set to EXPIRED.|

Reviewer 2: ```suggestion
| requestStatus | string | Enum {'REQUESTED', 'GRANTED', 'REVOKED', 'EXPIRED'}<br>When stating a request, this is set to REQUESTED. After response or update from master control set to GRANTED or REVOKED. If lease time expires set to EXPIRED.|
```

> | requestStatus | enum | Enum {'REQUESTED', 'GRANTED', 'REVOKED', 'EXPIRED'}<br>When stating a request, this is set to REQUESTED. After response or update from master control set to GRANTED or REVOKED. If lease time expires set to EXPIRED.|

Reviewer 2: ```suggestion
| requestStatus | string | Enum {'REQUESTED', 'GRANTED', 'REVOKED', 'EXPIRED'}<br>When stating a request, this is set to REQUESTED. After response or update from master control set to GRANTED or REVOKED. If lease time expires set to EXPIRED.|
```

## 6.13 Topic "visualization"

For a near real-time position and planned trajectory update the AGV can broadcast its position, velocity and planned trajectory on the topic `visualization`.

> The structure of the visualization object is the same as the position, velocity, planned path and intermediate path object in the state.

Reviewer 3: ```suggestion
The fields of the visualization object use the same structure as the position, velocity, planned path and intermediate path object in the state.
```

| maxLoadMass | float64 | [kg], Maximum loadable mass. |
| localizationTypes | array of string | Simplified description of localization type.<br/>Example values:<br/>NATURAL: natural landmarks,<br/>REFLECTOR: laser reflectors,<br/>RFID: RFID tags,<br/>DMC: data matrix code,<br/>SPOT: magnetic spots,<br/>GRID: magnetic grid.<br/>
| navigationTypes | array of string | Array of path planning types supported by the AGV, sorted by priority.<br/>Example values:<br/>PHYSICAL_LINE_GUIDED: no path planning, the AGV follows physical installed paths,<br/>VIRTUAL_LINE_GUIDED: the AGV follows fixed (virtual) paths,<br/>AUTONOMOUS: the AGV plans its path autonomously.|

> | *supportedZones* | array of string | Array of zone types supported by the vehicle.<br/>Enum {'BLOCKED', 'LINE-GUIDED', 'RELEASE', 'COORDINATED-REPLANNING', 'SPEED-LIMIT', 'ACTION'}.

Reviewer 3: Are DIRECTED/BIDIRECTED/PRIORITY/PENALTY missing here?
